---
title: "Orthogroup visualisation - gene annotations"
author: "Sofie Agerbæk"
date: "2025-10-14"
output: html_document
---

```{r setup, include=FALSE}
library(dplyr)
library(tidyverse)
library(ggplot2)
library(tidyr)
library(stringr)
library(cowplot)
library(pheatmap)
library(grid)
library(readxl)
```

```{r, include=FALSE}
options(digits = 3)        # number of digits printed by R default (vectors, data.frames, lists)
options(pillar.sigfig = 3) # number of digits printed by tibbles default.
options(width = 200)

text_base_size   <- 9 # in pt
fig.witdh        <- 180  # in mm
fig.height       <- 125  # in mm

# Set all text in plots to same size
theme_set(theme_cowplot(font_size = 10, rel_small = 1, rel_tiny = 1, rel_large = 1))
theme_update(panel.grid.major = element_line(colour="gray85"), panel.grid.minor = element_line(colour="gray95", linetype = 7))
# Setting output sizes for plots
knitr::opts_chunk$set(fig.width = fig.witdh/25.4)
knitr::opts_chunk$set(fig.height = fig.height/25.4)
knitr::opts_chunk$set(dpi = 108)

# Setting text size inside plots (geom_text, geom_label etc.)
ggplot_text_size <- text_base_size / ggplot2::.pt

options(ggplot2.continuous.fill  = scale_fill_viridis_c)
options(ggplot2.discrete.fill    = c("slategray2", "#404e67","#809bce","#9887D9" ,"#9C5BA2", "#DB9FF6",  "#d1625c", "#ffd6a5", "#e89c81"))
options(ggplot2.continuous.colour = scale_colour_viridis_c)
options(ggplot2.discrete.colour   = c("slategray2","#404e67", "#809bce","#9C5BA2","#9887D9",  "#d1625c", "#ffd6a5", "#e89c81"))

article_pallette <- c("#5674b4", "#7eacd1", "#afd9e9", "#e1f2f9", "#f7e18e", "#f7b060", "#ee6d43", "#d32e28")

```


##### Species order and taxonomy
```{r}
species_order <- c("Rhodotorula toruloides", "Cryptococcus neoformans", "Trichosporon ovoides",
                   "Malassezia furfur", "Malassezia equina", "Malassezia dermatis", "Malassezia sympodialis",
                   "Malassezia pachydermatis", "Malassezia restricta", "Malassezia globosa",
                   "Pichia kudriavzevii", "Saccharomyces cerevisiae", "Nakaseomyces glabratus",
                   "Candida metapsilosis", "Candida orthopsilosis", "Candida parapsilosis", "Candida tropicalis",
                   "Candida albicans", "Candida dubliniensis",
                   "Hortaea werneckii",  "Aspergillus fumigatus", "Histoplasma capsulatum",
                   "Chrysosporium queenslandicum",
                   "Microsporum canis", "Microsporum audouinii", "Microsporum ferrugineum",
                   "Nannizzia gypsea", "Epidermophyton floccosum", "Trichophyton schoenleinii",
                   "Trichophyton tonsurans", "Trichophyton mentagrophytes", "Trichophyton indotineae",
                   "Trichophyton interdigitale", "Trichophyton concentricum", "Trichophyton soudanese",
                   "Trichophyton kuryangei", "Trichophyton rubrum", "Trichophyton violaceum" )

species_order_hybrids <- c("Rhodotorula toruloides", "Cryptococcus neoformans", "Trichosporon ovoides",
                   "Trichosporon ovoides hybrid",
                   "Malassezia furfur", "Malassezia furfur hybrid", "Malassezia equina", "Malassezia dermatis",
                   "Malassezia sympodialis",
                   "Malassezia pachydermatis", "Malassezia restricta", "Malassezia globosa",
                   "Pichia kudriavzevii", "Saccharomyces cerevisiae", "Nakaseomyces glabratus",
                   "Candida metapsilosis", "Candida orthopsilosis", "Candida parapsilosis", "Candida tropicalis",
                   "Candida albicans", "Candida dubliniensis",
                   "Hortaea werneckii", "Hortaea werneckii hybrid", "Aspergillus fumigatus", "Histoplasma capsulatum",
                   "Chrysosporium queenslandicum",
                   "Microsporum canis", "Microsporum audouinii", "Microsporum ferrugineum",
                   "Nannizzia gypsea", "Epidermophyton floccosum", "Trichophyton schoenleinii",
                   "Trichophyton tonsurans", "Trichophyton mentagrophytes", "Trichophyton indotineae",
                   "Trichophyton interdigitale", "Trichophyton concentricum", "Trichophyton soudanese",
                   "Trichophyton kuryangei", "Trichophyton rubrum", "Trichophyton violaceum" )

```


```{r}
taxonomy <- data.frame(
  sample = c(
    "Aspergillus_fumigatus_IP24",
    "Candida_albicans_F133_05",
    "Candida_dubliniensis_11945",
    "Candida_dubliniensis_2747",
    "Candida_dubliniensis_8501",
    "Candida_metapsilosis_Y_48470",
    "Candida_orthopsilosis_Y_48468",
    "Candida_parapsilosis_11301",
    "Candida_parapsilosis_1954",
    "Candida_tropicalis_MYCT_0816",
    "Chrysosporium_queenslandicum_28077",
    "Cryptococcus_neoformans_KN99",
    "Epidermophyton_floccosum_100148",
    "Epidermophyton_floccosum_10867",
    "Epidermophyton_floccosum_130802",
    "Histoplasma_capsulatum_G186A",
    "Hortaea_werneckii_100455",
    "Hortaea_werneckii_1163",
    "Hortaea_werneckii_41051",
    "Malassezia_dermatis_9169",
    "Malassezia_equina_9969",
    "Malassezia_furfur_141402",
    "Malassezia_furfur_1878",
    "Malassezia_furfur_6000",
    "Malassezia_globosa_7705",
    "Malassezia_globosa_7886",
    "Malassezia_globosa_8744",
    "Malassezia_pachydermatis_21004",
    "Malassezia_restricta_7877",
    "Malassezia_restricta_7991",
    "Malassezia_sympodialis_44340",
    "Microsporum_audouinii_119449",
    "Microsporum_audouinii_33268",
    "Microsporum_canis_113480",
    "Microsporum_canis_15669",
    "Microsporum_canis_44551",
    "Microsporum_ferrugineum_131111",
    "Microsporum_ferrugineum_37371",
    "Nakaseomyces_glabratus_11311",
    "Nakaseomyces_glabratus_138",
    "Nakaseomyces_glabratus_15698",
    "Nannizzia_gypsea_17164",
    "Pichia_kudriavzevii_5732",
    "Rhodotorula_toruloides_Z1",
    "Saccharomyces_cerevisiae_wine010",
    "Trichophyton_concentricum_109405",
    "Trichophyton_concentricum_19626",
    "Trichophyton_indotineae_BE17",
    "Trichophyton_interdigitale_BE4",
    "Trichophyton_kuryangei_IHEM13979",
    "Trichophyton_mentagrophytes_124404",
    "Trichophyton_mentagrophytes_43573",
    "Trichophyton_rubrum_117539",
    "Trichophyton_rubrum_17065",
    "Trichophyton_rubrum_28986",
    "Trichophyton_rubrum_73588",
    "Trichophyton_schoenleinii_45859",
    "Trichophyton_schoenleinii_85571",
    "Trichophyton_soudanese_IHEM19743",
    "Trichophyton_tonsurans_112188",
    "Trichophyton_tonsurans_12935",
    "Trichophyton_tonsurans_72988",
    "Trichophyton_violaceum_73088",
    "Trichosporon_ovoides_4098",
    "Trichosporon_ovoides_7612",
    "Trichosporon_ovoides_9430"
  ),
  species = c(
    "Aspergillus_fumigatus" ,"Candida_albicans","Candida_dubliniensis"
    ,"Candida_dubliniensis", "Candida_dubliniensis","Candida_metapsilosis",
    "Candida_orthopsilosis","Candida_parapsilosis", "Candida_parapsilosis","Candida_tropicalis","Chrysosporium_queenslandicum",
    "Cryptococcus_neoformans","Epidermophyton_floccosum","Epidermophyton_floccosum",
    "Epidermophyton_floccosum","Histoplasma_capsulatum","Hortaea_werneckii" ,"Hortaea_werneckii","Hortaea_werneckii","Malassezia_dermatis","Malassezia_equina",
    "Malassezia_furfur","Malassezia_furfur","Malassezia_furfur","Malassezia_globosa",
    "Malassezia_globosa","Malassezia_globosa","Malassezia_pachydermatis",
    "Malassezia_restricta","Malassezia_restricta","Malassezia_sympodialis",
    "Microsporum_audouinii","Microsporum_audouinii","Microsporum_canis",
    "Microsporum_canis","Microsporum_canis","Microsporum_ferrugineum",
    "Microsporum_ferrugineum","Nakaseomyces_glabratus","Nakaseomyces_glabratus",
    "Nakaseomyces_glabratus", "Nannizzia_gypsea","Pichia_kudriavzevii",
    "Rhodotorula_toruloides","Saccharomyces_cerevisiae", "Trichophyton_concentricum",
    "Trichophyton_concentricum" ,   "Trichophyton_indotineae"  ,   "Trichophyton_interdigitale",
    "Trichophyton_kuryangei"   ,    "Trichophyton_mentagrophytes" , "Trichophyton_mentagrophytes",
    "Trichophyton_rubrum"      ,    "Trichophyton_rubrum"     ,    "Trichophyton_rubrum"    ,
    "Trichophyton_rubrum"     ,     "Trichophyton_schoenleinii"  ,  "Trichophyton_schoenleinii",
    "Trichophyton_soudanese"     ,  "Trichophyton_tonsurans"   ,   "Trichophyton_tonsurans"  ,
    "Trichophyton_tonsurans"    ,   "Trichophyton_violaceum"  ,     "Trichosporon_ovoides"   ,
    "Trichosporon_ovoides"       ,  "Trichosporon_ovoides"),
  genus = c(
    "Aspergillus",
    rep("Candida", 9),
    "Chrysosporium",
    "Cryptococcus",
    rep("Epidermophyton", 3),
    "Histoplasma",
    rep("Hortaea", 3),
    rep("Malassezia", 12),
    rep("Microsporum", 7),
    rep("Nakaseomyces", 3),
    "Nannizzia",
    "Pichia",
    "Rhodotorula",
    "Saccharomyces",
    rep("Trichophyton", 18),
    rep("Trichosporon", 3)
  ),
  family = c(
    "Trichocomaceae",     # Aspergillus
    rep("Debaryomycetaceae", 9),  # Candida spp.
    "Onygenaceae",        # Chrysosporium
    "Tremellaceae",       # Cryptococcus
    rep("Arthrodermataceae", 3),  # Epidermophyton
    "Ajellomycetaceae",   # Histoplasma
    rep("Teratosphaeriaceae", 3), # Hortaea
    rep("Malasseziaceae", 12),    # Malassezia spp.
    rep("Arthrodermataceae", 7),  # Microsporum spp.
    rep("Saccharomycetaceae", 3), # Nakaseomyces glabratus
    "Arthrodermataceae",  # Nannizzia
    "Saccharomycetaceae", # Pichia
    "Sporidiobolaceae",   # Rhodotorula
    "Saccharomycetaceae", # Saccharomyces
    rep("Arthrodermataceae", 18), # Trichophyton spp.
    rep("Trichosporonaceae", 3)   # Trichosporon spp.
  ),
  stringsAsFactors = FALSE
)

```



## Figure 4 - Visualising the distribution of EC and KEGG categories

### Data generation: KEGG pathway mapping

The following chunks extract EC numbers from gene annotations, map them to KEGG pathways via the KEGG API, and summarise gene counts per metabolic category per species. This involves slow API calls, so the results are saved to file.

```{r, eval=FALSE}
library(KEGGREST)

annotation_mat <- read.csv("../generated/all_annotations_combined.tsv")
annotation_mat$X <- NULL
head(annotation_mat, 3)

# 21.5 % of genes has EC annotations
sum(annotation_mat$EC != "") / length(annotation_mat$EC)

EC_df <- annotation_mat %>%
  filter(!EC == "")  %>%
  mutate(EC = sub("EC:", "", EC)) %>%
  select(species, strain_id, gene_id, gene_name, description, EC)

head(EC_df, 3)
```

```{r, eval=FALSE}
# Map EC to KEGG orthology (KO)
ko_ec <- keggLink("ko", "enzyme")

ko_ec <- data.frame(
  KO = sub("ko:", "", ko_ec),
  EC = sub("ec:", "", names(ko_ec)),
  stringsAsFactors = FALSE
)

annot_ko <- EC_df %>%
  left_join(ko_ec, by = c("EC" = "EC")) %>%
  filter(!is.na(KO))
```

Function to map KO to KEGG pathway. Batching into batches of 10 to limit server inquiries.
```{r, eval=FALSE}
get_kegg_info <- function(kos) {
  res <- list()

  # Split KOs into batches of 10
  for (i in seq(1, length(kos), by = 10)) {
    batch <- kos[i:min(i + 9, length(kos))]
    infos <- tryCatch(keggGet(batch), error = function(e) NULL)

    if (!is.null(infos)) {
      for (info in infos) {
        ko <- info$ENTRY
        pathways <- if (!is.null(info$PATHWAY)) names(info$PATHWAY) else NA
        category <- if (!is.null(info$CLASS)) {
          if (length(info$CLASS) >= 2) info$CLASS[2] else info$CLASS[1]
        } else NA
        res[[ko]] <- data.frame(
          KO = ko,
          Pathway = pathways,
          Category = category,
          stringsAsFactors = FALSE
        )
      }
    }
    Sys.sleep(0.2)  # polite delay between batches
  }

  bind_rows(res)
}

unique_kos <- unique(annot_ko$KO)
kegg_info <- get_kegg_info(unique_kos)
```

Get KEGG categories from pathway info
```{r, eval=FALSE}
pathway_cache <- new.env()

pathway_to_category <- function(map_ids) {
  res <- lapply(map_ids, function(id) {
    if (is.na(id) || id == "") return(NA)

    # Check cache first
    if (exists(id, envir = pathway_cache)) {
      return(get(id, envir = pathway_cache))
    }

    # Retrieve from KEGG
    info <- tryCatch(KEGGREST::keggGet(id)[[1]], error = function(e) NULL)

    val <- NA
    if (!is.null(info$CLASS)) {
      class_vec <- strsplit(info$CLASS, "; ")[[1]]

      # If this pathway belongs to "Metabolism", extract its subcategory
      if (length(class_vec) >= 2 && grepl("^Metabolism", class_vec[1], ignore.case = TRUE)) {
        val <- class_vec[2]  # e.g. "Carbohydrate metabolism"
      } else {
        val <- "Other"
      }
    }

    # Cache the result
    assign(id, val, envir = pathway_cache)
    val
  })

  names(res) <- map_ids
  unlist(res)
}

kegg_info$Category <- pathway_to_category(kegg_info$Pathway)
table(kegg_info$Category)
```

```{r, eval=FALSE}
annot_final <- annot_ko %>%
  left_join(kegg_info, by = "KO")

summary_df <- annot_final %>%
  filter(!is.na(Category)) %>%
  group_by(species, Category) %>%
  summarise(GeneCount = n_distinct(gene_id), .groups = "drop")

# write.csv(summary_df, "../generated/species_kegg_category_summary.csv", row.names = FALSE)
```


### Data generation: EC hydrolase subcategories

We focus on the hydrolases (EC = 3.x.x.x) as these are most interesting when it comes to metabolism.

```{r, eval=FALSE}
hydrolase_subcategory <- function(ec_nums) {
  sapply(ec_nums, function(ec) {
    if (is.na(ec) || ec == "") return(NA)
    parts <- strsplit(ec, "\\.")[[1]]

    # Only hydrolases
    if (parts[1] != "3") return(NA)

    second <- parts[2]
    switch(second,
           "1" = "Ester",
           "2" = "Glycosylases",
           "3" = "Ether",
           "4" = "Peptide",
           "5" = "Carbon–nitrogen",
           "6" = "Acid anhydrides",
           "7" = "Carbon–carbon",
           NA)  # ignore rare others
  })
}

EC_df$Hydrolase_sub <- hydrolase_subcategory(EC_df$EC)

hydro_summary <- EC_df %>%
  filter(!is.na(Hydrolase_sub)) %>%
  group_by(species, Hydrolase_sub) %>%
  summarise(GeneCount = n_distinct(gene_id), .groups = "drop") %>%
  group_by(species) %>%
  mutate(catprop = GeneCount / sum(GeneCount) * 100) %>%
  ungroup()

# write.csv(hydro_summary, "../generated/EC_hydrolase_summary.csv", row.names = FALSE)
```


### KEGG categories plot

```{r}
summary_df <- read.csv("../generated/species_kegg_category_summary.csv")
```

```{r}
ggplot(summary_df, aes(y = species, x = GeneCount, fill = Category)) +
  geom_bar(stat = "identity", position = "fill") +  # normalized proportion
  labs(y = "Proportion of genes", fill = "KEGG category") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

```

Plotting only selected categories
```{r}
selected_categories <- c("Amino acid metabolism", "Carbohydrate metabolism", "Energy metabolism",
                         "Lipid metabolism", "Nucleotide metabolism")

plot_summary_df <- summary_df %>%
  group_by(species) %>%
  mutate(catprop = GeneCount / sum(GeneCount) * 100) %>%
  ungroup() %>%
  filter(Category %in% selected_categories)

plot_summary_df$species <- sub("_", " ", plot_summary_df$species)
plot_summary_df$Category <- sub("metabolism", "", plot_summary_df$Category)

plot_summary_df$species <- factor(plot_summary_df$species, levels = rev(species_order))

KEGG_plot <- ggplot(plot_summary_df, aes(y = species, x = catprop, fill = Category)) +
  geom_bar(stat = "identity", position = "identity", alpha = 0.9) +
  theme_minimal() +
  theme(axis.ticks.x = element_blank(), panel.spacing = unit(1.3, "lines"), axis.title = element_blank(),
        panel.grid.minor = element_blank(), panel.grid.major.y = element_blank(),
        panel.grid.major.x = element_line(colour = "grey80", linetype = "dashed"),
        axis.text.y = element_text(face = "bold.italic", color = "black", size = 7),
        strip.background = element_blank(),
        strip.text = element_text(size = 8, face = "bold", color = "black")) +
  facet_wrap(~Category, nrow = 1, scales = "free_x")+
  guides(fill = 'none')
KEGG_plot

```

```{r}
# pdf("../plots/KEGG_plot.pdf", width = 6.5, height = 5)
# KEGG_plot
# dev.off()
```


### EC hydrolase plot

```{r}
hydro_summary <- read.csv("../generated/EC_hydrolase_summary.csv")
```

```{r}
ggplot(hydro_summary, aes(y = species, x = GeneCount, fill = Hydrolase_sub)) +
  geom_bar(stat = "identity", position = "fill") +  # normalized proportion
  labs(y = "Proportion of genes", fill = "EC hydrolase category") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

```

```{r}
selected_categories <- c("Ester", "Glycosylases", "Peptide", "Acid anhydrides", "Carbon–nitrogen")

plot_hydro_summary <- hydro_summary %>%
  group_by(species) %>%
  mutate(catprop = GeneCount / sum(GeneCount) * 100) %>%
  ungroup() %>%
  filter(Hydrolase_sub %in% selected_categories)

plot_hydro_summary$species <- sub("_", " ", plot_hydro_summary$species)

plot_hydro_summary$species <- factor(plot_hydro_summary$species, levels = rev(species_order))

EC_plot <- ggplot(plot_hydro_summary, aes(y = species, x = catprop, fill = Hydrolase_sub)) +
  geom_bar(stat = "identity", position = "identity", alpha = 0.9) +
  theme_minimal() +
  theme(axis.ticks.x = element_blank(), panel.spacing = unit(1.3, "lines"), axis.title = element_blank(),
        panel.grid.minor = element_blank(), panel.grid.major.y = element_blank(),
        panel.grid.major.x = element_line(colour = "grey80", linetype = "dashed"),
        axis.text.y = element_text(face = "bold.italic", color = "black", size = 7),
        strip.background = element_blank(),
        strip.text = element_text(size = 8, face = "bold", color = "black")) +
  facet_wrap(~Hydrolase_sub, nrow = 1, scales = "free_x")+
  guides(fill = 'none')
EC_plot
```

```{r}
# pdf("../plots/EC_plot.pdf", width = 6.5, height = 5)
# EC_plot
# dev.off()
```



## Figure 5 - Secreted protein metabolism heatmap


### Data generation: OG collapsing and annotation merging

```{r, eval=FALSE}
all_anno <- read.csv("../generated/all_annotations_combined.tsv")
all_anno$X <- NULL

secr_genes <- all_anno %>%
  filter(signalp == "True" & tmhmm == "0" & netgpi == "Not")

N0 <- read.csv("../results/Results_Aug23/Phylogenetic_Hierarchical_Orthogroups/N0.tsv", sep = "\t", header = TRUE)

N0 <- N0 %>%
  select(-c(HOG, Gene.Tree.Parent.Clade))

# Collapsing the OG rows and keeping just the unique gene ids
collapsed_df <- N0 %>%
  group_by(OG) %>%
  summarise(across(
    everything(),
    ~ {
      vals <- unlist(str_split(.x, ",\\s*"))
      vals <- vals[vals != "" & !is.na(vals)]
      paste(unique(vals), collapse = ",")
    },
     .names = "{.col}"
  ), .groups = "drop")

collapsed_long <- collapsed_df %>%
  pivot_longer(cols = -OG, names_to = c("sample"), values_to = "gene_ids") %>%
  filter(gene_ids != "") %>%
  mutate(gene_id = str_split(gene_ids, ",\\s*")) %>%
  select(-gene_ids) %>%
  unnest(cols = c(gene_id)) %>%
  mutate(
    gene_id = if_else(
    str_detect(gene_id, "\\|"),
    word(gene_id, 3, sep = "\\|"),
    gene_id)
  )

# Since orthofinder was run on fastas with slightly differently named headers,
# we match just the numbers (actual gene ids)
collapsed_long$gene_id <- sub("-T1", "", collapsed_long$gene_id)
collapsed_long$sample <- sub("\\.", "_", collapsed_long$sample)

collapsed_long <- collapsed_long %>%
  mutate(strain_id = map_chr(str_split(sample, pattern = "_"),
                           ~ paste(.x[3:length(.x)], collapse = "_")),
         species = map_chr(str_split(sample, pattern = "_"),
                           ~ paste(.x[1:2], collapse = "_")),
         gene_nr = as.character(str_extract_all(gene_id, "\\d+"))
  )

secr_genes$gene_nr <- as.character(str_extract_all(secr_genes$gene_id, "\\d+"))
all_anno$gene_nr   <- as.character(str_extract_all(all_anno$gene_id, "\\d+"))

# Old sample numbers (not CBS) are in the annotation df - these have to be fixed
old_nr_to_cbs <- c(
  "1" = "10867",
  "3" = "4098",
  "6" = "112188",
  "8" = "45859",
  "9" = "124404",
  "10" = "1163",
  "11" = "138",
  "12" = "1954",
  "13" = "15669",
  "27" = "7705",
  "28" = "7877",
  "29" = "141402",
  "31" = "5732",
  "32" = "28986",
  "34" = "2747",
  "410" = "41051",
  "332" = "33268",
  "445" = "44551",
  "171" = "17164",
  "435" = "43573",
  "735" = "73588",
  "855" = "85571",
  "129" = "12935",
  "739" = "72988",
  "730" = "73088")

all_anno_fix <- all_anno %>%
  mutate(strain_id = if_else(strain_id %in% names(old_nr_to_cbs),
                             old_nr_to_cbs[strain_id],
                             strain_id))

# Merge all anno so all the genes get any OG they might be a part of
collapsed_long$strain_id <- sub("_", "-", collapsed_long$strain_id)

merged_df <- merge(all_anno_fix, collapsed_long, by = c("species", "strain_id", "gene_nr"), all.x = TRUE)

annot_x_OGs <- merged_df

# write_tsv(merged_df, file = "../generated/all_annotations_x_OGs.tsv")

rm(merged_df, all_anno, collapsed_df, N0, secr_genes, old_nr_to_cbs)
```


### Data generation: OG summary and secreted protein tables

```{r, eval=FALSE}
annot_x_OGs <- read.csv("../generated/all_annotations_x_OGs.tsv", sep = "\t")

annot_x_OGs <- annot_x_OGs %>%
  select(1:2, OG, gene_id.y, everything())

# Pulling only the description columns. Then merging each OG for each sample
# into one row, where the most abundant not 'hypothetical protein' description is kept
merging_desc_df <- annot_x_OGs %>%
  select(!c(gene_id.x, antismash, source_file, sample)) %>%
  filter(!is.na(OG)) %>%
  mutate(IPR = sapply(str_extract_all(IPR, "IPR\\d{6}"), function(x) {
      if (length(x) == 0) NA else paste(unique(x), collapse = ";")
    }))

#renaming the COG so they match across samples - only the one-letter category
merging_desc_df <- merging_desc_df %>%
  mutate(COG = str_extract_all(COG, "(?<=\\b)[A-Z](?=\\b)")) %>%
  mutate(COG = sapply(COG, function(x) paste(unique(x), collapse = ";"))) %>%
  mutate(EC = sub("EC:", "", EC))

merging_desc_sum <- merging_desc_df %>%
  group_by(species, strain_id, OG) %>%
  summarise(
    description = ifelse(
      all(description == "hypothetical protein"),
      "hypothetical protein",
      names(sort(table(description[description != "hypothetical protein"]), decreasing = TRUE))[1]
    ),
    gene_id.y = paste(gene_id.y, collapse = "; "),
    gene_copies_in_OG = n(),
    IPR = paste(unique(IPR[!is.na(IPR)]), collapse = "; "),
    signalp = sum(signalp == "True"),
    tmhmm_geneswtmh = sum(tmhmm != "0"),
    netgpi_geneswithgpi = sum(netgpi != "Not"),
    COG = paste(unique(COG[!is.na(COG)]), collapse = "; "),
    EC = paste(unique(EC[!is.na(EC)]), collapse = "; "),
    PFAM = paste(unique(PFAM[!is.na(PFAM)]), collapse = "; ")
  )

merging_desc_sum <- merging_desc_sum %>%
  left_join(taxonomy %>% distinct(species, genus, family), by = "species")

OG_df <- merging_desc_sum %>%
  mutate(sample = paste(species, strain_id, sep = "_")) %>%
  pivot_wider(names_from = sample, values_from = gene_copies_in_OG, values_fill = 0) %>%
  group_by(OG) %>%
  summarise(
    number_of_species = n_distinct(species),
    description = ifelse(
      all(description == "hypothetical protein"),
      "hypothetical protein",
      names(sort(table(description[description != "hypothetical protein"]), decreasing = TRUE))[1]
    ),
    IPR = {
      terms <- unlist(str_split(IPR, "\\s*;\\s*"))
      terms <- terms[terms != "" & !is.na(terms)]
      terms <- unique(terms)
      if (length(terms) == 0) NA_character_ else paste(terms, collapse = ";")
    },
    signalp = sum(signalp, na.rm = TRUE),
    tmhmm_geneswtmh = sum(tmhmm_geneswtmh, na.rm = TRUE),
    netgpi_geneswithgpi = sum(netgpi_geneswithgpi),
    COG = {
      terms <- unlist(str_split(COG, "\\s*;\\s*"))
      terms <- terms[terms != "" & !is.na(terms)]
      terms <- unique(terms)
      if (length(terms) == 0) NA_character_ else paste(terms, collapse = ";")
    },
    EC = {
      terms <- unlist(str_split(EC, "\\s*;\\s*"))
      terms <- terms[terms != "" & !is.na(terms)]
      terms <- unique(terms)
      if (length(terms) == 0) NA_character_ else paste(terms, collapse = ";")
    },
    PFAM = {
      terms <- unlist(str_split(PFAM, "\\s*;\\s*"))
      terms <- terms[terms != "" & !is.na(terms)]
      terms <- unique(terms)
      if (length(terms) == 0) NA_character_ else paste(terms, collapse = ";")
    },
    number_of_genera = n_distinct(genus),
    number_of_families = n_distinct(family),
    across(all_of(names(.)[15:ncol(.)]), ~ sum(.x, na.rm = TRUE)),
    .groups = "drop"
  )

# write_tsv(OG_df, file = "../generated/OG_annotation_genecounts.tsv")

secr_df <- OG_df %>%
  filter(signalp > 0) %>%
  mutate(total_genec = rowSums(select(., 13:ncol(.)))) %>%
  select(OG,description,IPR,number_of_species,total_genec, COG, EC, PFAM, everything())

# write_tsv(secr_df, file = "../generated/nofilter_secr_OGxAnnos.tsv")
```


### Data generation: IPR enrichment for hypothetical proteins

```{r, eval=FALSE}
secr_df <- read.csv(file = "../generated/nofilter_secr_OGxAnnos.tsv", sep = "\t")

# Working on the hypothetical proteins that have IPR terms
hypo_w_ipr <- secr_df %>% filter(description == "hypothetical protein" & IPR != "")

ipr_map <- read.table("../data/ipr_entry_list.txt", sep = "\t", header = TRUE)

df_new <- hypo_w_ipr %>%
  # Split multiple IPR terms into separate rows
  mutate(IPR = str_split(IPR, ";")) %>%
  unnest(IPR) %>%
  # Join to InterPro table
  left_join(ipr_map %>% select(ENTRY_AC, ENTRY_NAME),
            by = c("IPR" = "ENTRY_AC")) %>%
  # Collapse back to OG-level description
  group_by(OG) %>%
  summarise(
    ipr_description = ifelse(sum(is.na(ENTRY_NAME)) == length(ENTRY_NAME),
                         "hypothetical protein",
                         paste(na.omit(unique(ENTRY_NAME)), collapse = "; ")),
    IPR.y =  paste(na.omit(unique(IPR)), collapse = "; "),
    .groups = "drop"
  )

merged <- merge(secr_df, df_new, by = "OG", all.x = TRUE)

merged <- merged %>%
  mutate(description = ifelse(description == "hypothetical protein" & ipr_description != "hypothetical protein"
                              & !is.na(ipr_description),
                               ipr_description,
                               description )) %>%
  select(-IPR.y, -ipr_description)

# write.csv(merged, file = "../generated/w_ipr_desc_nofilter_secr_OGxAnnos.tsv")
```


### Data generation: categorization by keywords, COG, EC, and PFAM

```{r, eval=FALSE}
cat_keywords_metabolism <- list(
  "carbohydrate metabolism" = c(
    "glyco", "glucosidase", "glycosidase", "glucanase", "chitinase", "mannosidase",
    "cellulase", "saccharid", "invertase", "amylase", "xylanase", "polysaccharide",
    "beta-glucosidase", "glycosyl", "glycosyltransferase", "glycoside hydrolase", "glycosylase", "glucose",
    "chitin", "chitinase", "deacetylase", "deacetylation", "CAZy", "GH", "glycoside hydrolase", "CBM", "LysM", "glucuronidase"
  ),
  "protein metabolism" = c(
    "protease", "peptidase", "aminopeptidase", "endopeptidase", "exopeptidase",
    "serine protease", "aspartyl", "metalloprotease", "cysteine protease", "peptidyl",
    "proteinase", "proteasome", "ubiquitin", "peptidyl-prolyl", "serine/threonine", "protein kinase"
  ),
  "lipid metabolism" = c(
    "lipase", "phospholipase", "acyltransferase", "acyl", "esterase", "fatty-acid",
    "lipoxygenase", "triacylglycerol", "glycerolipid", "lipid",
    "sterol", "cholesterol", "prenyl", "isoprenoid", "phosphatidyl", "sphingolipid", "ceramide", "lanosterol", "fatty", "SUR1"))

canonical_map <- list(
  "glucosidase" = c("glucosidase", "beta-glucosidase", "glycosidase", "glycoside hydrolase", "glycosylase", "glycohydrolase"),
  "chitinase" = c("chitinase"),
  "amylase" = c("amylase", "alpha-amylase"),
  "xylanase" = c("xylanase"),
  "cellulase" = c("cellulase"),
  "mannosidase" = c("mannosidase"),
  "aspartyl_protease" = c("aspartyl protease", "aspartyl", "sap", "secreted aspartyl protease"),
  "serine_protease" = c("serine protease"),
  "metalloprotease" = c("metalloprotease"),
  "peptidase" = c("peptidase", "protease", "proteinase", "aminopeptidase", "endopeptidase"),
  "lipase" = c("lipase", "phospholipase", "phospholipase b", "phospholipase a"),
  "adhesin" = c("adhesin", "agglutinin", "flocculin", "adhesion", "cell wall adhesin", "gpi-anchored adhesin"),
  "toxin" = c("toxin", "hemolysin", "invasin")
)

get_category <- function(description, cat_keywords) {
  description <- tolower(description)
  for (category in names(cat_keywords)) {
    patterns <- paste0(cat_keywords[[category]], collapse = "|")
    if (str_detect(description, patterns)) {
      return(category)
    }
  }
  return(NA)
}

get_canonical <- function(description, canonical_map) {
  description <- tolower(description)
  for (canon in names(canonical_map)) {
    patterns <- paste0(canonical_map[[canon]], collapse = "|")
    if (str_detect(description, patterns)) {
      return(canon)
    }
  }
  return(NA)
}

merged <- merged %>% filter(number_of_species > 1 & signalp > 3)

merged <- merged %>%
  mutate(category = sapply(description, get_category, cat_keywords = cat_keywords_metabolism))

merged <- merged %>%
  mutate(canonical = sapply(description, get_canonical, canonical_map = canonical_map))

# Getting more categories from COG
cog_map <- c(
  "C" = "energy metabolism",
  "G" = "carbohydrate metabolism",
  "E" = "protein metabolism",
  "I" = "lipid metabolism"
)

merged <- merged %>%
  mutate(category = ifelse(is.na(category) & !is.na(COG),
              paste(unique(na.omit(cog_map[strsplit(COG, ";")[[1]]])), collapse = "; "),
              category
  ))

# Getting more categories from EC
ec_map <- list(
  "carbohydrate metabolism" = c(
    "^1\\.1\\.", "^1\\.2\\.", "^2\\.4\\.", "^3\\.2\\.", "^4\\.2\\."
  ),
  "lipid metabolism" = c(
    "^3\\.1\\.1\\.", "^3\\.1\\.4\\.", "^2\\.3\\."
  ),
  "protein metabolism" = c(
    "^3\\.4\\.", "^2\\.7\\."
  ))

get_category_from_ec <- function(ec_number, ec_map) {
  if (is.na(ec_number)) return(NA)
  for (cat in names(ec_map)) {
    patterns <- ec_map[[cat]]
    if (any(str_detect(tolower(ec_number), tolower(patterns)))) {
      return(cat)
    }
  }
  return(NA)
}

merged <- merged %>%
  mutate(category = ifelse(
    is.na(category) & !is.na(EC),
    sapply(EC, get_category_from_ec, ec_map = ec_map),
    category
  ))

# Filter out OGs with too many transmembrane domains
merged <- merged %>%
  filter(!tmhmm_geneswtmh >= 0.75*signalp)

# write_tsv(merged, file = "../generated/formanual_cur.tsv")
```

More categories from PFAM
```{r, eval=FALSE}
test <- read_tsv("../generated/secreted_OGs_with_categories.tsv")

pfam_map <- read.delim("../generated/Pfam-A.clans.tsv", header = FALSE, sep = "\t")

pfam_lookup <- pfam_map[, c(1, 5)]
colnames(pfam_lookup) <- c("Pfam_ID", "Description")

replace_pfams <- function(pfam_string, lookup) {
  ids <- str_split(pfam_string, "[,;]")[[1]]
  names <- lookup[match(ids, lookup$Pfam_ID)]$Description
  names[is.na(names)] <- ids[is.na(names)]
  paste(unique(names), collapse = "; ")
}

test$pfam_names <- map_chr(test$PFAM, replace_pfams, lookup = pfam_lookup)

test2 <- test %>%
  mutate(category = ifelse(is.na(category),
                           sapply(pfam_names, get_category, cat_keywords = cat_keywords_metabolism),
                           category
                           ))

# write_tsv(test2, file = "../generated/formanual_cur.tsv")
```


### Metabolism heatmap from manually curated descriptions

```{r}
test_cur <- read_xlsx("../generated/1211_manual_cur_secreted.xlsx")
test_cur$netgpi_geneswithgpi <- as.numeric(test_cur$netgpi_geneswithgpi)
```


We collapse OGs with the same broad descriptions and get the gene copies as sums of all gene copies. This is kept at the sample level, so for each unique description we count how many gene copies each sample has.
```{r}
collapsed_df <- test_cur %>%
  filter(!category == "NA") %>%
  select(-c("description", "IPR", "PFAM", "EC", "COG", "canonical", "pfam_names")) %>%
  mutate(desc_collapse = str_trim(tolower(desc_collapse))) %>%
  group_by(desc_collapse) %>%
  summarise(OG = paste(unique(OG), collapse = "; "),
            category = first(category),
            number_of_species = max(number_of_species, na.rm = TRUE),
            total_genec = sum(total_genec, na.rm = TRUE),
            number_of_genera = max(number_of_genera, na.rm = TRUE),
            number_of_families = max(number_of_families, na.rm = TRUE),
            signalp = sum(as.numeric(signalp), na.rm = TRUE),
            tmhmm_geneswtmh = sum(as.numeric(tmhmm_geneswtmh), na.rm = TRUE),
            netgpi_geneswithgpi = sum(netgpi_geneswithgpi, na.rm = TRUE),
            across(all_of(names(.)[9:(ncol(.)-1)]), ~ sum(.x, na.rm = TRUE)),
            )

```


Creating the plot dataframes
```{r}
collapsed_filtered <- collapsed_df %>%
  filter(tmhmm_geneswtmh < 0.7*signalp) %>%
  filter(number_of_species > 2)

plot_df <- collapsed_filtered %>% select(desc_collapse, category, names(collapsed_df)[11:76])

plot_df_long <- pivot_longer(plot_df, cols = -c(desc_collapse, category), names_to = "sample", values_to = "gene_copies")

plot_df_long <- merge(plot_df_long, taxonomy[1:2], by = "sample")

# Handling hybrid samples
plot_df_long[plot_df_long$sample == "Malassezia_furfur_1878",]$species <- "Malassezia furfur hybrid"
plot_df_long[plot_df_long$sample == "Malassezia_furfur_6000",]$species <- "Malassezia furfur hybrid"
plot_df_long[plot_df_long$sample == "Trichosporon_ovoides_9430",]$species <- "Trichosporon ovoides hybrid"
plot_df_long[plot_df_long$sample ==  "Hortaea_werneckii_41051",]$species <- "Hortaea werneckii hybrid"
plot_df_long[plot_df_long$sample ==  "Hortaea_werneckii_100455",]$species <- "Hortaea werneckii hybrid"

plot_df_long <- plot_df_long %>%
  select(-sample) %>%
  group_by(desc_collapse, species) %>%
  summarise(gene_copies = round(mean(gene_copies)), # Average for species with multiple samples
            category = first(category))

plot_df_n <- pivot_wider(plot_df_long, names_from = species, values_from = gene_copies)

plot_df_n$desc_collapse <- str_to_title(plot_df_n$desc_collapse)
plot_df <- column_to_rownames(plot_df_n, "desc_collapse")

category_order <- c(
  "carbohydrate metabolism",
  "lipid metabolism",
  "protein metabolism",
  "virulence associated",
  "adhesin-associated"
)

plot_df$category <- factor(plot_df$category, levels = category_order)
plot_df <- plot_df %>% arrange(category)

```


Plotting the metabolism heatmap
```{r}
subset_cats <- c("carbohydrate metabolism", "lipid metabolism", "protein metabolism")

names(plot_df)[2:42] <- sub("_", " ", names(plot_df)[2:42])

# Filter and order the dataframe
plot_df_sub <- plot_df %>%
  filter(category %in% subset_cats) %>%
  mutate(category = factor(category, levels = category_order)) %>%
  arrange(category)

# numeric matrix
num_df <- plot_df_sub[, species_order_hybrids] %>%
  mutate(across(everything(), as.numeric))
mat <- as.matrix(num_df)
rownames(mat) <- rownames(plot_df_sub)

# Z scores for colour scale
mat_z <- t(scale(t(mat)))  # center and scale each row
rownames(mat_z) <- rownames(mat)
mat_z[mat_z > 2] <- 2
mat_z[mat_z < -2] <- -2

# Build annotation
annotation_row <- data.frame(Category = plot_df_sub$category)
rownames(annotation_row) <- rownames(mat)

# Compute white gaps
gaps_row <- cumsum(table(plot_df_sub$category))
gaps_row <- gaps_row[1:3]

# Numeric labels
num_labels <- matrix(round(mat, 1), nrow = nrow(mat), ncol = ncol(mat))
rownames(num_labels) <- rownames(mat)
colnames(num_labels) <- colnames(mat)

# Plot
plot <- pheatmap(mat_z,
         cluster_rows = FALSE,
         cluster_cols = FALSE,
         show_rownames = TRUE,
         show_colnames = TRUE,
         fontsize_row = 10,
         fontsize_col = 12,
         display_numbers = num_labels,
         number_color = "black",
         fontsize_number = 8,
         gaps_row = gaps_row,
         border_color = "white",
         color = article_pallette,
         angle_col = 90,
         legend = FALSE,
         cellheight=16,
         cellwidth=15)

plot
```

Bold text formatting
```{r}
gt <- plot$gtable

# Modify the graphical parameters of species and gene product labels
gt$grobs[[2]]$gp <- gpar(fontface = "bold.italic", fontsize = 12)
gt$grobs[[3]]$gp <- gpar(fontface = "bold", fontsize = 10)

grid.newpage()
grid.draw(gt)
```

```{r}
# pdf("../plots/1311_secreted_metabolites.pdf", width = 11, height = 12, useDingbats = TRUE)
# grid.draw(gt)
# dev.off()

# png("../plots/1311_secreted_metabolites.png", width = 11, height = 10.5, units = "in", res = 300)
# grid.draw(gt)
# dev.off()
```



## Figure 6 - GPI anchored + virulence/adhesin

### Data generation: GPI-anchored protein selection

```{r, eval=FALSE}
secr_df <- read.csv(file = "../generated/w_ipr_desc_nofilter_secr_OGxAnnos.tsv")

gpi_df <- secr_df %>% filter(netgpi_geneswithgpi > 0) %>% select(-X)

# write.csv(gpi_df, "../generated/gpi_for_manual_cur.csv")
```


### GPI-anchored heatmap from manually curated data

```{r}
gpi_df <- read_xlsx("../generated/gpi_manual_curated.xlsx")
gpi_df <- gpi_df %>% dplyr::select(!"Column1")

```


```{r}
collapsed_gpi <- gpi_df %>%
  dplyr::select(-c("description", "IPR", "PFAM", "EC", "COG")) %>%
  group_by(desc_collapse) %>%
  summarise(OG = paste(unique(OG), collapse = "; "),
            number_of_species = max(number_of_species, na.rm = TRUE),
            total_genec = sum(total_genec, na.rm = TRUE),
            number_of_genera = max(number_of_genera, na.rm = TRUE),
            number_of_families = max(number_of_families, na.rm = TRUE),
            signalp = sum(as.numeric(signalp), na.rm = TRUE),
            tmhmm_geneswtmh = sum(as.numeric(tmhmm_geneswtmh), na.rm = TRUE),
            netgpi_geneswithgpi = sum(netgpi_geneswithgpi, na.rm = TRUE),
            across(all_of(names(.)[10:(ncol(.))]), ~ sum(.x, na.rm = TRUE)),
            )

```


```{r}
gpi_plot_df <- collapsed_gpi %>% dplyr::select(desc_collapse, names(collapsed_gpi)[10:75])

gpi_plot_df_long <- pivot_longer(gpi_plot_df, cols = -desc_collapse, names_to = "sample", values_to = "gene_copies")

gpi_plot_df_long$sample <- sub("\\.", "-", gpi_plot_df_long$sample)

gpi_plot_df_long <- merge(gpi_plot_df_long, taxonomy[1:2], by = "sample")

# Handling hybrid samples
gpi_plot_df_long[gpi_plot_df_long$sample == "Malassezia_furfur_1878",]$species <- "Malassezia furfur hybrid"
gpi_plot_df_long[gpi_plot_df_long$sample == "Malassezia_furfur_6000",]$species <- "Malassezia furfur hybrid"
gpi_plot_df_long[gpi_plot_df_long$sample == "Trichosporon_ovoides_9430",]$species <- "Trichosporon ovoides hybrid"
gpi_plot_df_long[gpi_plot_df_long$sample ==  "Hortaea_werneckii_41051",]$species <- "Hortaea werneckii hybrid"
gpi_plot_df_long[gpi_plot_df_long$sample ==  "Hortaea_werneckii_100455",]$species <- "Hortaea werneckii hybrid"

gpi_plot_df_long <- gpi_plot_df_long %>%
  select(-sample) %>%
  group_by(desc_collapse, species) %>%
  summarise(gene_copies = round(mean(gene_copies))) # Average for species with multiple samples

gpi_plot_df_long_filtered <- gpi_plot_df_long %>%
  group_by(desc_collapse) %>%
  filter(sum(gene_copies != 0) > 1) %>%
  ungroup()

gpi_plot_df_n <- pivot_wider(gpi_plot_df_long_filtered, names_from = species, values_from = gene_copies)

gpi_plot_df_n <- gpi_plot_df_n %>%
  arrange(desc_collapse != "GPI-anchored, unknown function", desc_collapse)

gpi_plot_df <- column_to_rownames(gpi_plot_df_n, "desc_collapse")

```


Plotting GPI-anchored heatmap
```{r}
names(gpi_plot_df)[1:41] <- sub("_", " ", names(gpi_plot_df)[1:41])

# numeric matrix
num_df <- gpi_plot_df[, species_order_hybrids] %>%
  mutate(across(everything(), as.numeric))
mat <- as.matrix(num_df)
rownames(mat) <- rownames(gpi_plot_df)

rns_ordered <- c(setdiff(sort(rownames(mat)), "GPI-anchored, unknown function"),
                 "GPI-anchored, unknown function")

# reorder matrix rows
mat <- mat[rns_ordered, ]

# Z scores for colour scale
mat_z <- t(scale(t(mat)))  # center and scale each row
rownames(mat_z) <- rownames(mat)
mat_z[mat_z > 2] <- 2
mat_z[mat_z < -2] <- -2

# Numeric labels
num_labels <- matrix(round(mat, 1), nrow = nrow(mat), ncol = ncol(mat))
rownames(num_labels) <- rownames(mat)
colnames(num_labels) <- colnames(mat)

# Plot
plot <- pheatmap(mat_z,
         cluster_rows = FALSE,
         cluster_cols = FALSE,
         show_rownames = TRUE,
         show_colnames = TRUE,
         fontsize_row = 7,
         fontsize_col = 10,
         display_numbers = num_labels,
         number_color = "black",
         fontsize_number = 7,
         color = article_pallette,
         angle_col = 90,
         legend = FALSE)

plot

```

Bold text formatting
```{r}
gt <- plot$gtable

# Modify the graphical parameters of species and gene product labels
gt$grobs[[2]]$gp <- gpar(fontface = "bold.italic") #species labels
gt$grobs[[3]]$gp <- gpar(fontface = "bold", fontsize = 8) #descriptions

grid.newpage()
grid.draw(gt)

```

```{r}
# pdf("../plots/0511_gpi_anchored.pdf", width = 10, height = 10, useDingbats = TRUE)
# grid.draw(gt)
# dev.off()

# png("../plots/0411_secreted_metabolites.png", width = 10, height = 10, units = "in", res = 300)
# grid.draw(gt)
# dev.off()
```


### Virulence/adhesin heatmap

Adhesin and virulence associated from manually curated excel sheets - both secreted and non-secreted.
```{r}
test_cur <- read_xlsx("../generated/1211_manual_cur_secreted.xlsx")
virulence_assoc <- test_cur %>% filter(category == "virulence associated"| category == "adhesin-associated") %>%
  dplyr::select(-c("description", "IPR", "PFAM", "EC", "COG", "canonical"))

viru_non_secr <- read_xlsx("../generated/1211_manual_cur_secreted.xlsx", sheet = 2) %>%
  filter(!OG %in% test_cur$OG) %>% dplyr::select(-c("description","IPR", "PFAM", "EC", "COG"))

names(viru_non_secr)[9:75] <- sub("\\.", "-", names(viru_non_secr)[9:75])

viru_non_secr$total_genec <- rowSums(viru_non_secr[9:74] != 0)

viru_non_secr$secr <- FALSE
virulence_assoc$secr <- TRUE

comb_viru <- rbind(viru_non_secr, virulence_assoc)

table(comb_viru$category)
```


```{r}
collapsed_gpi_viru <- comb_viru %>%
  group_by(desc_collapse) %>%
  dplyr::select(-pfam_names) %>%
  filter(!is.na(category)) %>%
  summarise(OG = paste(unique(OG), collapse = "; "),
            number_of_species = max(number_of_species, na.rm = TRUE),
            total_genec = sum(total_genec, na.rm = TRUE),
            number_of_genera = max(number_of_genera, na.rm = TRUE),
            number_of_families = max(number_of_families, na.rm = TRUE),
            signalp = sum(as.numeric(signalp), na.rm = TRUE),
            tmhmm_geneswtmh = sum(as.numeric(tmhmm_geneswtmh), na.rm = TRUE),
            netgpi_geneswithgpi = sum(as.numeric(netgpi_geneswithgpi), na.rm = TRUE),
            category = first(category),
            secr = if (length(unique(secr)) == 1) as.character(unique(secr)) else "mixed",
            across(all_of(names(.)[9:(ncol(.)-3)]), ~ sum(.x, na.rm = TRUE)),
            )

```


```{r}
gpi_plot_df <- collapsed_gpi_viru %>% dplyr::select(desc_collapse, category, secr, netgpi_geneswithgpi, names(collapsed_gpi_viru)[12:77]) %>% filter(!is.na(category))

gpi_plot_df_long <- pivot_longer(gpi_plot_df, cols = -c(desc_collapse, category, secr, netgpi_geneswithgpi), names_to = "sample", values_to = "gene_copies")

gpi_plot_df_long <- merge(gpi_plot_df_long, taxonomy[1:2], by = "sample")

# Handling hybrid samples
gpi_plot_df_long[gpi_plot_df_long$sample == "Malassezia_furfur_1878",]$species <- "Malassezia furfur hybrid"
gpi_plot_df_long[gpi_plot_df_long$sample == "Malassezia_furfur_6000",]$species <- "Malassezia furfur hybrid"
gpi_plot_df_long[gpi_plot_df_long$sample == "Trichosporon_ovoides_9430",]$species <- "Trichosporon ovoides hybrid"
gpi_plot_df_long[gpi_plot_df_long$sample ==  "Hortaea_werneckii_41051",]$species <- "Hortaea werneckii hybrid"
gpi_plot_df_long[gpi_plot_df_long$sample ==  "Hortaea_werneckii_100455",]$species <- "Hortaea werneckii hybrid"

gpi_plot_df_long <- gpi_plot_df_long %>%
  select(-sample) %>%
  group_by(desc_collapse, species) %>%
  summarise(gene_copies = round(mean(gene_copies)), # Average for species with multiple samples
            category = first(category),
            secr = first(secr),
            GPI = first(netgpi_geneswithgpi))

gpi_plot_df_long_filtered <- gpi_plot_df_long %>%
  group_by(desc_collapse) %>%
  filter(sum(gene_copies != 0) > 1) %>%
  ungroup()

gpi_plot_df_n <- pivot_wider(gpi_plot_df_long_filtered, names_from = species, values_from = gene_copies)

gpi_plot_df_n <- gpi_plot_df_n[order(
  gpi_plot_df_n$category,
  gpi_plot_df_n$desc_collapse == "GPI-anchored, unannotated"
), ]

gpi_plot_df <- column_to_rownames(gpi_plot_df_n, "desc_collapse")

```


Plotting virulence/adhesin heatmap
```{r}
names(gpi_plot_df)[2:44] <- sub("_", " ", names(gpi_plot_df)[2:44])

# numeric matrix
num_df <- gpi_plot_df[, species_order_hybrids] %>%
  mutate(across(everything(), as.numeric))
mat <- as.matrix(num_df)
rownames(mat) <- rownames(gpi_plot_df)

# Z scores for colour scale
mat_z <- t(scale(t(mat)))  # center and scale each row
rownames(mat_z) <- rownames(mat)
mat_z[mat_z > 2] <- 2
mat_z[mat_z < -2] <- -2

# Build annotation
annotation_row <- data.frame(GPI = as.character(gpi_plot_df$GPI>0), secreted = gpi_plot_df$secr)
rownames(annotation_row) <-  (rownames(mat))

# Compute white gaps
gaps_row <- cumsum(table(gpi_plot_df$category))

# Numeric labels
num_labels <- matrix(round(mat, 1), nrow = nrow(mat), ncol = ncol(mat))
rownames(num_labels) <- rownames(mat)
colnames(num_labels) <- colnames(mat)

# Plot
plot <- pheatmap(mat_z,
         cluster_rows = FALSE,
         cluster_cols = FALSE,
         annotation_row = annotation_row,
         show_rownames = TRUE,
         show_colnames = TRUE,
         fontsize_row = 8,
         fontsize_col = 10,
         display_numbers = num_labels,
         number_color = "black",
         fontsize_number = 8,
         gaps_row = gaps_row,
         border_color = "white",
         color = article_pallette,
         angle_col = 90,
         legend = FALSE,
       cellheight=10,
       cellwidth=13)

plot

```

Bold text formatting
```{r}
gt <- plot$gtable

# Modify the graphical parameters of species and gene product labels
gt$grobs[[2]]$gp <- gpar(fontface = "bold.italic", fontsize = 10) #species labels
gt$grobs[[3]]$gp <- gpar(fontface = "bold", fontsize = 9) #descriptions

grid.newpage()
grid.draw(gt)

```

```{r}
# pdf("../plots/2111_gpi_viru_hm.pdf", width = 10.5, height = 11, useDingbats = TRUE)
# grid.draw(gt)
# dev.off()

# png("../plots/2111_gpi_viru_hm.png", width = 10.5, height = 11, units = "in", res = 300)
# grid.draw(gt)
# dev.off()
```



## Supplementary: OG annotation table with PFAM names and categories

This section generates the complete OG annotation table with PFAM descriptions and functional categories.

```{r, eval=FALSE}
OGs_x_annos <- read.csv("../generated/OG_annotation_genecounts.tsv", sep = "\t")

pfam_map <- read.delim("../generated/Pfam-A.clans.tsv", header = FALSE, sep = "\t")

pfam_lookup <- pfam_map[, c(1, 5)]
colnames(pfam_lookup) <- c("Pfam_ID", "Description")

replace_pfams <- function(pfam_string, lookup) {
  ids <- str_split(pfam_string, "[,;]")[[1]]
  names <- lookup$Description[match(ids, lookup$Pfam_ID)]
  names[is.na(names)] <- ids[is.na(names)]
  paste(unique(names), collapse = "; ")
}

OGs_x_annos$pfam_names <- map_chr(OGs_x_annos$PFAM, replace_pfams, lookup = pfam_lookup)

cat_keywords <- list(
  "carbohydrate metabolism" = c(
    "glyco", "glucosidase", "glycosidase", "glucanase", "chitinase", "mannosidase",
    "cellulase", "saccharid", "invertase", "amylase", "xylanase", "polysaccharide",
    "beta-glucosidase", "glycosyl", "glycosyltransferase", "glycoside hydrolase", "glycosylase", "glucose",
    "chitin", "chitinase", "deacetylase", "deacetylation", "CAZy", "GH", "glycoside hydrolase", "CBM", "LysM", "glucuronidase"
  ),
  "protein metabolism" = c(
    "protease", "peptidase", "aminopeptidase", "endopeptidase", "exopeptidase",
    "serine protease", "aspartyl", "metalloprotease", "cysteine protease", "peptidyl",
    "proteinase", "proteasome", "ubiquitin", "peptidyl-prolyl", "serine/threonine", "protein kinase"
  ),
  "lipid metabolism" = c(
    "lipase", "phospholipase", "acyltransferase", "acyl", "esterase", "fatty-acid",
    "lipoxygenase", "triacylglycerol", "glycerolipid", "lipid",
    "sterol", "cholesterol", "prenyl", "isoprenoid", "phosphatidyl", "sphingolipid", "ceramide", "lanosterol", "fatty", "SUR1"
    ),
 "virulence associated" = c(
    "toxin", "virulence", "pathogen", "pathogenic", "invasin", "virulence factor",
    "hemolysin", "phospholipase b", "secreted aspartyl protease", "sap", "secreted protease",
    "lysM", "Hydrophobin", "Heat shock", "CFEM", "BNR repeat", "PIR", "necrosis", "stress", "igE"
  ),
  "adhesin-associated" = c(
    "adhesin", "adhesion", "agglutinin", "flocculin", "intimin", "fimbrial", "adhesive",
    "gpi-anchored adhesin", "gpi-anchored", "cell wall adhesin", "Flo11", "FLO8", "GLEYA",
    "^PA14", "Hyphal_reg", "^Hyr", "GPI", "Glycosylphosphatidylinositol"
  )
)

get_category <- function(description, cat_keywords) {
  description <- tolower(description)
  for (category in names(cat_keywords)) {
    patterns <- paste0(cat_keywords[[category]], collapse = "|")
    if (str_detect(description, patterns)) {
      return(category)
    }
  }
  return(NA)
}

OGs_x_annos <- OGs_x_annos %>%
  mutate(category = sapply(description, get_category, cat_keywords = cat_keywords)) %>%
  mutate(category = ifelse(
    is.na(category),
    sapply(pfam_names, get_category, cat_keywords = cat_keywords),
    category
  ))

# write_tsv(OGs_x_annos, "../generated/OGs_x_anno_incl_pfamnames.tsv")
```
