---
title: "Orthogroup visualisation - gene annotations"
author: "Sofie Agerbæk"
date: "2025-10-14"
output: html_document
---

```{r setup, include=FALSE}
library(dplyr)
library(tidyverse)
library(ggplot2)
library(plotly)
library(RColorBrewer)
library(heatmaply)
library(writexl)
library(Polychrome)
library(tidyr)
library(stringr)
```

```{r, include=FALSE}
library(cowplot)
options(digits = 3)        # number of digits printed by R default (vectors, data.frames, lists)
options(pillar.sigfig = 3) # number of digits printed by tibbles default.
options(width = 200)

text_base_size   <- 9 # in pt
fig.witdh        <- 180  # in mm
fig.height       <- 125  # in mm

# Set all text in plots to same size
theme_set(theme_cowplot(font_size = 10, rel_small = 1, rel_tiny = 1, rel_large = 1))
theme_update(panel.grid.major = element_line(colour="gray85"), panel.grid.minor = element_line(colour="gray95", linetype = 7))
# Setting output sizes for plots
knitr::opts_chunk$set(fig.width = fig.witdh/25.4)
knitr::opts_chunk$set(fig.height = fig.height/25.4)
knitr::opts_chunk$set(dpi = 108) # You need to find your monitors dpi yourself.

# Setting text size inside plots (geom_text, geom_label etc.)
ggplot_text_size <- text_base_size / ggplot2::.pt
# Now use: geom_text(..., size = ggplot_text_size)


options(ggplot2.continuous.fill  = scale_fill_viridis_c)
options(ggplot2.discrete.fill    = c("slategray2", "#404e67","#809bce","#9887D9" ,"#9C5BA2", "#DB9FF6",  "#d1625c", "#ffd6a5", "#e89c81"))
options(ggplot2.continuous.colour = scale_colour_viridis_c)
options(ggplot2.discrete.colour   = c("slategray2","#404e67", "#809bce","#9C5BA2","#9887D9",  "#d1625c", "#ffd6a5", "#e89c81"))

#A good palette for many distinct colours:
#dist_palette25 <- c("#86A3D8", "#F4C283", "#856697","#AA7A86", "#ACA9CD", "#EF6E77", "#DD9983", "#B016FC", "#E1C753", "#9D2E98", "#AEB951", "#4D85FB", "#D768D2",  "#6C947E", "#B04F76","#968FED", "#B04FB7",  "#312464", "#A03C18","#FB66AD", "#912EC5",  "#381CC9",  "#ED9552", "#A0A373", "#BC3D63")

pall1 <- c("#242D3B", "#404e67", "#809bce","#89B2E7", "#7fa9a7", "#7eb77f", "#a8cea9", "#d3ecc4", "#ecffc3", "#d5da97","#AA9F40", "#A27E31","#de8f49", "#D67238", "#C8461E", "#9B2603", "#420602", "#802F2F", "#C88080","#E694BC","#CE70B6" , "#9C5BA2", "#7442A3",  "#9887d0", "#6c6b9c", "#443654")

#pall1_light <- c("#6585BD", "#76A1F1","#9DC8FF", "#A3E1EF", "#96FBEA", "#8AD1B6", "#7DCF7E", "#B0EE86", "#ecffc3", "#d5da97","#E9F466","#DBD40F", "#AA9F40", "#A27E31", "#D67238","#de8f49", "#FF8A67", "#FF7066", "#E44242", "#D66161", "#C88080","#d2698c","#CE70B6","#E694BC","#FFC3E8", "#DB9FF6", "#987EBD",  "#8185E7", "#ABA9D6", "#979797", "#86A3D8", "#F4C283", "#856697","#AA7A86", "#ACA9CD", "#EF6E77", "#DD9983", "#B016FC")

pall1_light <- c("#979797","#A6AAB1", "#869BBE", "#6585BD", "#76A1F1","#9DC8FF", "#A3E1EF", "#96FBEA", "#8CDBD1",  "#84D09A", "#7DCF7E", "#B0EE86", "#CEF7A5", "#DCFFC2", "#F2FFA8","#F4FF7B", "#F2F853", "#DDDB63", "#C7BB4E", "#BF9E43", "#CB883E", "#E78332", "#EA6C3D", "#DD5A3D", "#E44242", "#D65050",  "#E27373", "#F27D7C", "#F99A99", "#FFB7B5", "#E990A1","#D1768E" ,"#C9799F", "#BF89B1", "#B599C4","#ABA9D6",  "#987EBD", "#9B8EAC")

#"#9887D9",
pall2_12 <- c("#404e67","#809bce", "slategray2", "#ABA9D6" , "#9C5BA2" , "#C88080", "#9B2603", "#e89c81", "#ffd6a5", "#d5da97", "#7eb77f",  "#7fa9a7")

# heatmap palette 

hm_pallette <- rev(c("#BA2C13", "#C85A36", "#DD9F6A", "#F0E49D", "#EAEFF2", "#C7D3E0", "#A4B7CE", "#7893B7"))

article_pallette <- c("#5674b4", "#7eacd1", "#afd9e9", "#e1f2f9", "#f7e18e", "#f7b060", "#ee6d43", "#d32e28")

```



##### Species order and taxonomy df 
```{r}
species_order <- c("Rhodotorula toruloides", "Cryptococcus neoformans", "Trichosporon ovoides",
                   "Malassezia furfur", "Malassezia equina", "Malassezia dermatis", "Malassezia sympodialis", 
                   "Malassezia pachydermatis", "Malassezia restricta", "Malassezia globosa", 
                   "Pichia kudriavzevii", "Saccharomyces cerevisiae", "Nakaseomyces glabratus",
                   "Candida metapsilosis", "Candida orthopsilosis", "Candida parapsilosis", "Candida tropicalis", 
                   "Candida albicans", "Candida dubliniensis",
                   "Hortaea werneckii",  "Aspergillus fumigatus", "Histoplasma capsulatum", 
                   "Chrysosporium queenslandicum", 
                   "Microsporum canis", "Microsporum audouinii", "Microsporum ferrugineum", 
                   "Nannizzia gypsea", "Epidermophyton floccosum", "Trichophyton schoenleinii",
                   "Trichophyton tonsurans", "Trichophyton mentagrophytes", "Trichophyton indotineae",
                   "Trichophyton interdigitale", "Trichophyton concentricum", "Trichophyton soudanese",
                   "Trichophyton kuryangei", "Trichophyton rubrum", "Trichophyton violaceum" )

```


```{r}
taxonomy <- data.frame(
  sample = c(
    "Aspergillus_fumigatus_IP24",
    "Candida_albicans_F133_05",
    "Candida_dubliniensis_11945",
    "Candida_dubliniensis_2747",
    "Candida_dubliniensis_8501",
    "Candida_metapsilosis_Y_48470",
    "Candida_orthopsilosis_Y_48468",
    "Candida_parapsilosis_11301",
    "Candida_parapsilosis_1954",
    "Candida_tropicalis_MYCT_0816",
    "Chrysosporium_queenslandicum_28077",
    "Cryptococcus_neoformans_KN99",
    "Epidermophyton_floccosum_100148",
    "Epidermophyton_floccosum_10867",
    "Epidermophyton_floccosum_130802",
    "Histoplasma_capsulatum_G186A",
    "Hortaea_werneckii_100455",
    "Hortaea_werneckii_1163",
    "Hortaea_werneckii_41051",
    "Malassezia_dermatis_9169",
    "Malassezia_equina_9969",
    "Malassezia_furfur_141402",
    "Malassezia_furfur_1878",
    "Malassezia_furfur_6000",
    "Malassezia_globosa_7705",
    "Malassezia_globosa_7886",
    "Malassezia_globosa_8744",
    "Malassezia_pachydermatis_21004",
    "Malassezia_restricta_7877",
    "Malassezia_restricta_7991",
    "Malassezia_sympodialis_44340",
    "Microsporum_audouinii_119449",
    "Microsporum_audouinii_33268",
    "Microsporum_canis_113480",
    "Microsporum_canis_15669",
    "Microsporum_canis_44551",
    "Microsporum_ferrugineum_131111",
    "Microsporum_ferrugineum_37371",
    "Nakaseomyces_glabratus_11311",
    "Nakaseomyces_glabratus_138",
    "Nakaseomyces_glabratus_15698",
    "Nannizzia_gypsea_17164",
    "Pichia_kudriavzevii_5732",
    "Rhodotorula_toruloides_Z1",
    "Saccharomyces_cerevisiae_wine010",
    "Trichophyton_concentricum_109405",
    "Trichophyton_concentricum_19626",
    "Trichophyton_indotineae_BE17",
    "Trichophyton_interdigitale_BE4",
    "Trichophyton_kuryangei_IHEM13979",
    "Trichophyton_mentagrophytes_124404",
    "Trichophyton_mentagrophytes_43573",
    "Trichophyton_rubrum_117539",
    "Trichophyton_rubrum_17065",
    "Trichophyton_rubrum_28986",
    "Trichophyton_rubrum_73588",
    "Trichophyton_schoenleinii_45859",
    "Trichophyton_schoenleinii_85571",
    "Trichophyton_soudanese_IHEM19743",
    "Trichophyton_tonsurans_112188",
    "Trichophyton_tonsurans_12935",
    "Trichophyton_tonsurans_72988",
    "Trichophyton_violaceum_73088",
    "Trichosporon_ovoides_4098",
    "Trichosporon_ovoides_7612",
    "Trichosporon_ovoides_9430"
  ),
  species = c(
    "Aspergillus_fumigatus" ,"Candida_albicans","Candida_dubliniensis"
    ,"Candida_dubliniensis", "Candida_dubliniensis","Candida_metapsilosis",
    "Candida_orthopsilosis","Candida_parapsilosis", "Candida_parapsilosis","Candida_tropicalis","Chrysosporium_queenslandicum",
    "Cryptococcus_neoformans","Epidermophyton_floccosum","Epidermophyton_floccosum",
    "Epidermophyton_floccosum","Histoplasma_capsulatum","Hortaea_werneckii" ,"Hortaea_werneckii","Hortaea_werneckii","Malassezia_dermatis","Malassezia_equina",
    "Malassezia_furfur","Malassezia_furfur","Malassezia_furfur","Malassezia_globosa",
    "Malassezia_globosa","Malassezia_globosa","Malassezia_pachydermatis",
    "Malassezia_restricta","Malassezia_restricta","Malassezia_sympodialis",
    "Microsporum_audouinii","Microsporum_audouinii","Microsporum_canis",
    "Microsporum_canis","Microsporum_canis","Microsporum_ferrugineum",
    "Microsporum_ferrugineum","Nakaseomyces_glabratus","Nakaseomyces_glabratus",
    "Nakaseomyces_glabratus", "Nannizzia_gypsea","Pichia_kudriavzevii",
    "Rhodotorula_toruloides","Saccharomyces_cerevisiae", "Trichophyton_concentricum",
    "Trichophyton_concentricum" ,   "Trichophyton_indotineae"  ,   "Trichophyton_interdigitale",
    "Trichophyton_kuryangei"   ,    "Trichophyton_mentagrophytes" , "Trichophyton_mentagrophytes", 
    "Trichophyton_rubrum"      ,    "Trichophyton_rubrum"     ,    "Trichophyton_rubrum"    ,   
    "Trichophyton_rubrum"     ,     "Trichophyton_schoenleinii"  ,  "Trichophyton_schoenleinii",
    "Trichophyton_soudanese"     ,  "Trichophyton_tonsurans"   ,   "Trichophyton_tonsurans"  , 
    "Trichophyton_tonsurans"    ,   "Trichophyton_violaceum"  ,     "Trichosporon_ovoides"   ,    
    "Trichosporon_ovoides"       ,  "Trichosporon_ovoides"),
  genus = c(
    "Aspergillus",
    rep("Candida", 9),
    "Chrysosporium",
    "Cryptococcus",
    rep("Epidermophyton", 3),
    "Histoplasma",
    rep("Hortaea", 3),
    rep("Malassezia", 12),
    rep("Microsporum", 7),
    rep("Nakaseomyces", 3),
    "Nannizzia",
    "Pichia",
    "Rhodotorula",
    "Saccharomyces",
    rep("Trichophyton", 18),
    rep("Trichosporon", 3)
  ),
  family = c(
    "Trichocomaceae",     # Aspergillus
    rep("Debaryomycetaceae", 9),  # Candida spp.
    "Onygenaceae",        # Chrysosporium
    "Tremellaceae",       # Cryptococcus
    rep("Arthrodermataceae", 3),  # Epidermophyton
    "Ajellomycetaceae",   # Histoplasma
    rep("Teratosphaeriaceae", 3), # Hortaea
    rep("Malasseziaceae", 12),    # Malassezia spp.
    rep("Arthrodermataceae", 7),  # Microsporum spp.
    rep("Saccharomycetaceae", 3), # Nakaseomyces glabratus
    "Arthrodermataceae",  # Nannizzia
    "Saccharomycetaceae", # Pichia
    "Sporidiobolaceae",   # Rhodotorula
    "Saccharomycetaceae", # Saccharomyces
    rep("Arthrodermataceae", 18), # Trichophyton spp.
    rep("Trichosporonaceae", 3)   # Trichosporon spp.
  ),
  stringsAsFactors = FALSE
)


```



## Figure 4 - Visualising the distribution of EC and KEGG categories 

### loading data 

```{r}
annotation_mat <- read.csv("../generated/all_annotations_combined.tsv")
annotation_mat$X <- NULL
head(annotation_mat, 3)
```



### Extract EC numbers and map to KEGG (KO) and KEGG pathways

```{r}
sum(annotation_mat$EC != "") / length(annotation_mat$EC)
# 21.5 % of genes has EC annotations 

EC_df <- annotation_mat %>%
  filter(!EC == "")  %>%
  mutate(EC = sub("EC:", "", EC)) %>%
  select(species, strain_id, gene_id, gene_name, description, EC)

head(EC_df, 3)
```

Map to KEGG (KO)
```{r}
ko_ec <- keggLink("ko", "enzyme")  # returns KO -> EC

ko_ec <- data.frame(
  KO = sub("ko:", "", ko_ec),
  EC = sub("ec:", "", names(ko_ec)),
  stringsAsFactors = FALSE
)

```


```{r}
annot_ko <- EC_df %>%
  left_join(ko_ec, by = c("EC" = "EC")) %>%
  filter(!is.na(KO))

```

Function to map KO to KEGG pathway (which we want for visualization)
I am batching into batches of 10, such that the amount of inquiries for the server is limited. 
```{r}
get_kegg_info <- function(kos) {
  res <- list()
  
  # Split KOs into batches of 10
  for (i in seq(1, length(kos), by = 10)) {
    batch <- kos[i:min(i + 9, length(kos))]
    infos <- tryCatch(keggGet(batch), error = function(e) NULL)
    
    if (!is.null(infos)) {
      for (info in infos) {
        ko <- info$ENTRY
        pathways <- if (!is.null(info$PATHWAY)) names(info$PATHWAY) else NA
        category <- if (!is.null(info$CLASS)) {
          if (length(info$CLASS) >= 2) info$CLASS[2] else info$CLASS[1]
        } else NA
        res[[ko]] <- data.frame(
          KO = ko,
          Pathway = pathways,
          Category = category,
          stringsAsFactors = FALSE
        )
      }
    }
    Sys.sleep(0.2)  # polite delay between batches 
  }
  
  bind_rows(res)
}

```


Get unique KOs and retrieve KEGG info
```{r}
unique_kos <- unique(annot_ko$KO)
kegg_info <- get_kegg_info(unique_kos)
```



Get KEGG categories from pathway info 

```{r}
pathway_cache <- new.env()

pathway_to_category <- function(map_ids) {
  res <- lapply(map_ids, function(id) {
    if (is.na(id) || id == "") return(NA)
    
    # Check cache first
    if (exists(id, envir = pathway_cache)) {
      return(get(id, envir = pathway_cache))
    }
    
    # Retrieve from KEGG
    info <- tryCatch(KEGGREST::keggGet(id)[[1]], error = function(e) NULL)
    
    val <- NA
    if (!is.null(info$CLASS)) {
      class_vec <- strsplit(info$CLASS, "; ")[[1]]
      
      # If this pathway belongs to "Metabolism", extract its subcategory
      if (length(class_vec) >= 2 && grepl("^Metabolism", class_vec[1], ignore.case = TRUE)) {
        val <- class_vec[2]  # e.g. "Carbohydrate metabolism"
      } else {
        val <- "Other"
      }
    }
    
    # Cache the result
    assign(id, val, envir = pathway_cache)
    val
  })
  
  names(res) <- map_ids
  unlist(res)
}

```

```{r}
kegg_info$Category <- pathway_to_category(kegg_info$Pathway)

# Check result
table(kegg_info$Category)
```


```{r}
annot_final <- annot_ko %>%
  left_join(kegg_info, by = "KO")
```


```{r}
#how many genes actually have EC / KEGG annos
sum(!is.na(annot_final$Pathway)) / length(annot_final$Pathway)
# 86.9% of genes with EC have KEGG pathway annotations 
sum(!is.na(annot_final$Category)) / length(annot_final$Category)
# 82.7 % of KEGG pathways get a category
sum(annot_final$Category == "Other", na.rm = TRUE) / sum(!is.na(annot_final$Category))
# 94.7 % of the categories for pathways are not metabolism  

summary_df <- annot_final %>%
  filter(!is.na(Category)) %>%
  group_by(species, Category) %>%
  summarise(GeneCount = n_distinct(gene_id), .groups = "drop")

```

write to file for easy later use and visu

```{r}
write.csv(summary_df, "../generated/species_kegg_category_summary.csv", row.names = FALSE)
```

 
### Read from file and plot

```{r}
summary_df <- read.csv("../generated/species_kegg_category_summary.csv")
```


plot
```{r}
ggplot(summary_df, aes(y = species, x = GeneCount, fill = Category)) +
  geom_bar(stat = "identity", position = "fill") +  # normalized proportion
  labs(y = "Proportion of genes", fill = "KEGG category") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

```

Plotting only selected categories
```{r}
selected_categories <- c("Amino acid metabolism", "Carbohydrate metabolism", "Energy metabolism", 
                         "Lipid metabolism", "Nucleotide metabolism")

plot_summary_df <- summary_df %>%
  group_by(species) %>%
  mutate(catprop = GeneCount / sum(GeneCount) * 100) %>%
  ungroup() %>%
  filter(Category %in% selected_categories) 

plot_summary_df$species <- sub("_", " ", plot_summary_df$species)
plot_summary_df$Category <- sub("metabolism", "", plot_summary_df$Category)

plot_summary_df$species <- factor(plot_summary_df$species, levels = rev(species_order))

KEGG_plot <- ggplot(plot_summary_df, aes(y = species, x = catprop, fill = Category)) +
  geom_bar(stat = "identity", position = "identity", alpha = 0.9) +  
  theme_minimal() +
  theme(axis.ticks.x = element_blank(), panel.spacing = unit(1.3, "lines"), axis.title = element_blank(),
        panel.grid.minor = element_blank(), panel.grid.major.y = element_blank(),
        panel.grid.major.x = element_line(colour = "grey80", linetype = "dashed"),
        axis.text.y = element_text(face = "bold.italic", color = "black", size = 7),
        strip.background = element_blank(),
        strip.text = element_text(size = 8, face = "bold", color = "black")) +
  facet_wrap(~Category, nrow = 1, scales = "free_x")+
#  expand_limits(x = c(NA, 15))+
  guides(fill = 'none')
KEGG_plot

```

save the plot to file
```{r}
pdf("../plots/KEGG_plot.pdf", width = 6.5, height = 5)
KEGG_plot
dev.off()

```


```{r}
rm(summary_df, annot_ko, annot_final, ko_ec, kegg_info, plot_summary_df, KEGG_plot)
```


### Plotting the distribution of EC annotations

We focus on the hydrolases (EC = 3.4.x.x) as these are most interesting when it comes to metabolism 

```{r}
hydrolase_subcategory <- function(ec_nums) {
  sapply(ec_nums, function(ec) {
    if (is.na(ec) || ec == "") return(NA)
    parts <- strsplit(ec, "\\.")[[1]]
    
    # Only hydrolases
    if (parts[1] != "3") return(NA)
    
    second <- parts[2]
    switch(second,
           "1" = "Ester",
           "2" = "Glycosylases",
           "3" = "Ether", 
           "4" = "Peptide",
           "5" = "Carbon–nitrogen",
           "6" = "Acid anhydrides",
           "7" = "Carbon–carbon",
           NA)  # ignore rare others
  })
}
```

```{r}
EC_df$Hydrolase_sub <- hydrolase_subcategory(EC_df$EC)

table(EC_df$Hydrolase_sub)

```

```{r}
hydro_summary <- EC_df %>%
  filter(!is.na(Hydrolase_sub)) %>%
  group_by(species, Hydrolase_sub) %>%
  summarise(GeneCount = n_distinct(gene_id), .groups = "drop") %>%
  group_by(species) %>%
  mutate(catprop = GeneCount / sum(GeneCount) * 100) %>%
  ungroup()
```

save to file for later 
```{r}
write.csv(hydro_summary, "../generated/EC_hydrolase_summary.csv", row.names = FALSE)
```

```{r}
hydro_summary <- read.csv("../generated/EC_hydrolase_summary.csv")

```


```{r}
ggplot(hydro_summary, aes(y = species, x = GeneCount, fill = Hydrolase_sub)) +
  geom_bar(stat = "identity", position = "fill") +  # normalized proportion
  labs(y = "Proportion of genes", fill = "EC hydrolase category") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

```

```{r}
selected_categories <- c("Ester", "Glycosylases", "Peptide", "Acid anhydrides", "Carbon–nitrogen")

plot_hydro_summary <- hydro_summary %>%
  group_by(species) %>%
  mutate(catprop = GeneCount / sum(GeneCount) * 100) %>%
  ungroup() %>%
  filter(Hydrolase_sub %in% selected_categories) 

plot_hydro_summary$species <- sub("_", " ", plot_hydro_summary$species)

plot_hydro_summary$species <- factor(plot_hydro_summary$species, levels = rev(species_order))

EC_plot <- ggplot(plot_hydro_summary, aes(y = species, x = catprop, fill = Hydrolase_sub)) +
  geom_bar(stat = "identity", position = "identity", alpha = 0.9) +  
  theme_minimal() +
  theme(axis.ticks.x = element_blank(), panel.spacing = unit(1.3, "lines"), axis.title = element_blank(),
        panel.grid.minor = element_blank(), panel.grid.major.y = element_blank(),
        panel.grid.major.x = element_line(colour = "grey80", linetype = "dashed"),
        axis.text.y = element_text(face = "bold.italic", color = "black", size = 7),
        strip.background = element_blank(),
        strip.text = element_text(size = 8, face = "bold", color = "black")) +
  facet_wrap(~Hydrolase_sub, nrow = 1, scales = "free_x")+
 # expand_limits(x = c(NA, 35))+
  guides(fill = 'none')
EC_plot
```

save plot 
```{r}
pdf("../plots/EC_plot.pdf", width = 6.5, height = 5)
EC_plot
dev.off()
```

```{r}
rm()
```



## Figure 5 

### Preparing dataframe from annotations - this can be moved to own script

#### Merging annotations with signalp/tmhmm/netgpi
```{r}

all_anno <- read.csv("../generated/all_annotations_combined.tsv")
all_anno$X <- NULL

all_anno$gene_id <- sub("-T1", "", all_anno$gene_id)

signalp <- read.csv("../generated/signalp_tmhmm_netgpi_annotation_matrix.tsv", sep = "\t", header = TRUE)
signalp$gene_id <- sub("-T1", "", signalp$gene_id)
signalp$strain_id <- sub("^0+", "", signalp$strain_id) #removing leading 0s as these are not in all_anno
```

Merging these two matrices
```{r}
all_anno <- all_anno %>%
  select(-c("signalp", "tmhmm", "netgpi"))

merged_anno <- merge(all_anno, signalp, by = c("species", "strain_id", "gene_id"), all.x = TRUE)

table(merged_anno[merged_anno$signalp == "True",]$species)
sum(merged_anno$signalp == "True", na.rm = TRUE)

missing_from_merged <- signalp %>%
  anti_join(merged_anno, by = c("species", "strain_id", "gene_id"))

nrow(missing_from_merged)
head(missing_from_merged)

table(missing_from_merged[1:2])
table(all_anno[1:2])

write.csv(merged_anno, file = "../generated/all_annotations_combined.tsv", sep = "\t")
```

#### Reading annotation from file and merging with OGs
```{r}
all_anno <- read.csv("../generated/all_annotations_combined.tsv")
all_anno$X <- NULL
```

```{r}
secr_genes <- all_anno %>%
  filter(signalp == "True" & tmhmm == "0" & netgpi == "Not")

```

```{r}
N0 <- read.csv("../results/Results_Aug23/Phylogenetic_Hierarchical_Orthogroups/N0.tsv", sep = "\t", header = TRUE)
head(N0[1:5])

N0 <- N0 %>%
  select(-c(HOG, Gene.Tree.Parent.Clade)) 

```

Collapsing the OG rows and keeping just the unique gene ids
```{r}
collapsed_df <- N0 %>%
  group_by(OG) %>%
  summarise(across(
    everything(),
    ~ {
      vals <- unlist(str_split(.x, ",\\s*"))
      vals <- vals[vals != "" & !is.na(vals)]
      paste(unique(vals), collapse = ",")
    },
     .names = "{.col}"
  ), .groups = "drop")


```

```{r}
collapsed_long <- collapsed_df %>%
  pivot_longer(cols = -OG, names_to = c("sample"), values_to = "gene_ids") %>%
  filter(gene_ids != "") %>%
  mutate(gene_id = str_split(gene_ids, ",\\s*")) %>%
  select(-gene_ids) %>%
  unnest(cols = c(gene_id)) %>%
  mutate(
    gene_id = if_else(
    str_detect(gene_id, "\\|"),
    word(gene_id, 3, sep = "\\|"),
    gene_id)
  )
  
```

Since orthofinder was run on fastas with slightly differently named headers (unfortunate), we wanna match just the numbers (actual gene ids)

```{r}
collapsed_long$gene_id <- sub("-T1", "", collapsed_long$gene_id) # to match with secreted df 
collapsed_long$sample <- sub("\\.", "_", collapsed_long$sample) 


collapsed_long <- collapsed_long %>%
  mutate(strain_id = map_chr(str_split(sample, pattern = "_"), 
                           ~ paste(.x[3:length(.x)], collapse = "_")), 
         species = map_chr(str_split(sample, pattern = "_"),
                           ~ paste(.x[1:2], collapse = "_")),
         gene_nr = as.character(str_extract_all(gene_id, "\\d+"))
  )

secr_genes$gene_nr <- as.character(str_extract_all(secr_genes$gene_id, "\\d+"))
all_anno$gene_nr   <- as.character(str_extract_all(all_anno$gene_id, "\\d+"))
```

Old sample numbers (not CBS) are in the annotation df - these have to be fixed
```{r}
old_nr_to_cbs <- c(
  "1" = "10867",
  "3" = "4098", 
  "6" = "112188",
  "8" = "45859",
  "9" = "124404",
  "10" = "1163",
  "11" = "138",
  "12" = "1954",
  "13" = "15669",
  "27" = "7705",
  "28" = "7877",
  "29" = "141402",
  "31" = "5732",
  "32" = "28986",
  "34" = "2747",
  "410" = "41051",
  "332" = "33268",
  "445" = "44551",
  "171" = "17164", 
  "435" = "43573",
  "735" = "73588",
  "855" = "85571",
  "129" = "12935",
  "739" = "72988",
  "730" = "73088")

names(old_nr_to_cbs)
missing_strain_ids_in_collapsed <- (unique(all_anno$strain_id)[!unique(all_anno$strain_id) %in% unique(collapsed_long$strain_id)])

missing_strain_ids_in_collapsed[!missing_strain_ids_in_collapsed %in% names(old_nr_to_cbs)]

all_anno_fix <- all_anno %>%
  mutate(strain_id = if_else(strain_id %in% names(old_nr_to_cbs),
                             old_nr_to_cbs[strain_id],
                             strain_id))

(unique(collapsed_long$strain_id)[!unique(collapsed_long$strain_id) %in% unique(all_anno_fix$strain_id)])
(unique(all_anno_fix$strain_id)[!unique(all_anno_fix$strain_id) %in% unique(collapsed_long$strain_id)])

```


Now i want to merge the two dataframes by specie, strain_id and gene_nr
We merge all anno, so all the genes get any OG they might be a part of (secreted or not)
```{r}
str(all_anno[c("species", "strain_id", "gene_nr")])
str(collapsed_long[c("species", "strain_id", "gene_nr")])

collapsed_long$strain_id <- sub("_", "-", collapsed_long$strain_id)

merged_df <- merge(all_anno_fix, collapsed_long, by = c("species", "strain_id", "gene_nr"), all.x = TRUE)

length(unique(merged_df$OG))
length(unique(collapsed_long$OG))


setdiff(unique(collapsed_long$OG), unique(merged_df$OG))

missing_from_merged <- collapsed_long %>%
  anti_join(merged_df, by = c("species", "strain_id", "gene_nr"))

nrow(missing_from_merged)
head(missing_from_merged)

#empty :)

annot_x_OGs <- merged_df

# unique(missing_from_merged$strain_id)
# unique(collapsed_long$strain_id)
# unique(all_anno$strain_id)
# 
# (unique(collapsed_long$strain_id)[!unique(collapsed_long$strain_id) %in% unique(all_anno$strain_id)])
# (unique(all_anno$strain_id)[!unique(all_anno$strain_id) %in% unique(collapsed_long$strain_id)])
```

write to tsv
```{r}
write_tsv(merged_df, file = "../generated/all_annotations_x_OGs.tsv")
```
 
Removing intermediary dataframes etc
```{r}
rm(merged_df, all_anno, collapsed_df, N0, secr_genes, old_nr_to_cbs, missing_from_merged)
```


#### Reading merged annotations x OG table from file 
```{r}
annot_x_OGs <- read.csv("../generated/all_annotations_x_OGs.tsv", sep = "\t") 

annot_x_OGs <- annot_x_OGs %>%
  select(1:2, OG, gene_id.y, everything())
```



Taxa look-up data frame  
```{r}
taxonomy <- data.frame(
  sample = c(
    "Aspergillus_fumigatus_IP24",
    "Candida_albicans_F133_05",
    "Candida_dubliniensis_11945",
    "Candida_dubliniensis_2747",
    "Candida_dubliniensis_8501",
    "Candida_metapsilosis_Y_48470",
    "Candida_orthopsilosis_Y_48468",
    "Candida_parapsilosis_11301",
    "Candida_parapsilosis_1954",
    "Candida_tropicalis_MYCT_0816",
    "Chrysosporium_queenslandicum_28077",
    "Cryptococcus_neoformans_KN99",
    "Epidermophyton_floccosum_100148",
    "Epidermophyton_floccosum_10867",
    "Epidermophyton_floccosum_130802",
    "Histoplasma_capsulatum_G186A",
    "Hortaea_werneckii_100455",
    "Hortaea_werneckii_1163",
    "Hortaea_werneckii_41051",
    "Malassezia_dermatis_9169",
    "Malassezia_equina_9969",
    "Malassezia_furfur_141402",
    "Malassezia_furfur_1878",
    "Malassezia_furfur_6000",
    "Malassezia_globosa_7705",
    "Malassezia_globosa_7886",
    "Malassezia_globosa_8744",
    "Malassezia_pachydermatis_21004",
    "Malassezia_restricta_7877",
    "Malassezia_restricta_7991",
    "Malassezia_sympodialis_44340",
    "Microsporum_audouinii_119449",
    "Microsporum_audouinii_33268",
    "Microsporum_canis_113480",
    "Microsporum_canis_15669",
    "Microsporum_canis_44551",
    "Microsporum_ferrugineum_131111",
    "Microsporum_ferrugineum_37371",
    "Nakaseomyces_glabratus_11311",
    "Nakaseomyces_glabratus_138",
    "Nakaseomyces_glabratus_15698",
    "Nannizzia_gypsea_17164",
    "Pichia_kudriavzevii_5732",
    "Rhodotorula_toruloides_Z1",
    "Saccharomyces_cerevisiae_wine010",
    "Trichophyton_concentricum_109405",
    "Trichophyton_concentricum_19626",
    "Trichophyton_indotineae_BE17",
    "Trichophyton_interdigitale_BE4",
    "Trichophyton_kuryangei_IHEM13979",
    "Trichophyton_mentagrophytes_124404",
    "Trichophyton_mentagrophytes_43573",
    "Trichophyton_rubrum_117539",
    "Trichophyton_rubrum_17065",
    "Trichophyton_rubrum_28986",
    "Trichophyton_rubrum_73588",
    "Trichophyton_schoenleinii_45859",
    "Trichophyton_schoenleinii_85571",
    "Trichophyton_soudanese_IHEM19743",
    "Trichophyton_tonsurans_112188",
    "Trichophyton_tonsurans_12935",
    "Trichophyton_tonsurans_72988",
    "Trichophyton_violaceum_73088",
    "Trichosporon_ovoides_4098",
    "Trichosporon_ovoides_7612",
    "Trichosporon_ovoides_9430"
  ),
  species = c(
    "Aspergillus_fumigatus" ,"Candida_albicans","Candida_dubliniensis"
    ,"Candida_dubliniensis", "Candida_dubliniensis","Candida_metapsilosis",
    "Candida_orthopsilosis","Candida_parapsilosis", "Candida_parapsilosis","Candida_tropicalis","Chrysosporium_queenslandicum",
    "Cryptococcus_neoformans","Epidermophyton_floccosum","Epidermophyton_floccosum",
    "Epidermophyton_floccosum","Histoplasma_capsulatum","Hortaea_werneckii" ,"Hortaea_werneckii","Hortaea_werneckii","Malassezia_dermatis","Malassezia_equina",
    "Malassezia_furfur","Malassezia_furfur","Malassezia_furfur","Malassezia_globosa",
    "Malassezia_globosa","Malassezia_globosa","Malassezia_pachydermatis",
    "Malassezia_restricta","Malassezia_restricta","Malassezia_sympodialis",
    "Microsporum_audouinii","Microsporum_audouinii","Microsporum_canis",
    "Microsporum_canis","Microsporum_canis","Microsporum_ferrugineum",
    "Microsporum_ferrugineum","Nakaseomyces_glabratus","Nakaseomyces_glabratus",
    "Nakaseomyces_glabratus", "Nannizzia_gypsea","Pichia_kudriavzevii",
    "Rhodotorula_toruloides","Saccharomyces_cerevisiae", "Trichophyton_concentricum",
    "Trichophyton_concentricum" ,   "Trichophyton_indotineae"  ,   "Trichophyton_interdigitale",
    "Trichophyton_kuryangei"   ,    "Trichophyton_mentagrophytes" , "Trichophyton_mentagrophytes", 
    "Trichophyton_rubrum"      ,    "Trichophyton_rubrum"     ,    "Trichophyton_rubrum"    ,   
    "Trichophyton_rubrum"     ,     "Trichophyton_schoenleinii"  ,  "Trichophyton_schoenleinii",
    "Trichophyton_soudanese"     ,  "Trichophyton_tonsurans"   ,   "Trichophyton_tonsurans"  , 
    "Trichophyton_tonsurans"    ,   "Trichophyton_violaceum"  ,     "Trichosporon_ovoides"   ,    
    "Trichosporon_ovoides"       ,  "Trichosporon_ovoides"),
  genus = c(
    "Aspergillus",
    rep("Candida", 9),
    "Chrysosporium",
    "Cryptococcus",
    rep("Epidermophyton", 3),
    "Histoplasma",
    rep("Hortaea", 3),
    rep("Malassezia", 12),
    rep("Microsporum", 7),
    rep("Nakaseomyces", 3),
    "Nannizzia",
    "Pichia",
    "Rhodotorula",
    "Saccharomyces",
    rep("Trichophyton", 18),
    rep("Trichosporon", 3)
  ),
  family = c(
    "Trichocomaceae",     # Aspergillus
    rep("Debaryomycetaceae", 9),  # Candida spp.
    "Onygenaceae",        # Chrysosporium
    "Tremellaceae",       # Cryptococcus
    rep("Arthrodermataceae", 3),  # Epidermophyton
    "Ajellomycetaceae",   # Histoplasma
    rep("Teratosphaeriaceae", 3), # Hortaea
    rep("Malasseziaceae", 12),    # Malassezia spp.
    rep("Arthrodermataceae", 7),  # Microsporum spp.
    rep("Saccharomycetaceae", 3), # Nakaseomyces glabratus
    "Arthrodermataceae",  # Nannizzia
    "Saccharomycetaceae", # Pichia
    "Sporidiobolaceae",   # Rhodotorula
    "Saccharomycetaceae", # Saccharomyces
    rep("Arthrodermataceae", 18), # Trichophyton spp.
    rep("Trichosporonaceae", 3)   # Trichosporon spp.
  ),
  stringsAsFactors = FALSE
)

family_dict <- c(
  "Aspergillus"     = "Aspergillaceae",
  "Candida"         = "Saccharomycetaceae",
  "Chrysosporium"   = "Onygenaceae",
  "Cryptococcus"    = "Cryptococcaceae",
  "Epidermophyton"  = "Arthrodermataceae",
  "Histoplasma"     = "Ajellomycetaceae",
  "Hortaea"         = "Teratosphaeriaceae",
  "Malassezia"      = "Malasseziaceae",
  "Microsporum"     = "Arthrodermataceae",
  "Nakaseomyces"    = "Saccharomycetaceae",
  "Nannizzia"       = "Arthrodermataceae",
  "Pichia"          = "Saccharomycetaceae",
  "Rhodotorula"     = "Sporidiobolaceae",
  "Saccharomyces"   = "Saccharomycetaceae",
  "Trichophyton"    = "Arthrodermataceae",
  "Trichosporon"    = "Trichosporonaceae"
)

```


Pulling only the description columns. Then merging each OG for each sample into one row, where the most abundant not 'hypothetical protein' description is kept
```{r}
merging_desc_df <- annot_x_OGs %>%
  select(!c(gene_id.x, antismash, source_file, sample)) %>%
  filter(!is.na(OG)) %>%
  mutate(IPR = sapply(str_extract_all(IPR, "IPR\\d{6}"), function(x) {
      if (length(x) == 0) NA else paste(unique(x), collapse = ";")
    }))

table(merging_desc_df$COG)

#renaming the COG so they match across samples - only the one-letter category 
merging_desc_df <- merging_desc_df %>%
  mutate(COG = str_extract_all(COG, "(?<=\\b)[A-Z](?=\\b)")) %>%
  mutate(COG = sapply(COG, function(x) paste(unique(x), collapse = ";"))) %>%
  mutate(EC = sub("EC:", "", EC))



```


```{r}
merging_desc_sum <- merging_desc_df %>%
  group_by(species, strain_id, OG) %>%
  summarise(
    description = ifelse(
      all(description == "hypothetical protein"),
      "hypothetical protein",
      names(sort(table(description[description != "hypothetical protein"]), decreasing = TRUE))[1]
    ),
    gene_id.y = paste(gene_id.y, collapse = "; "),
    gene_copies_in_OG = n(),
    IPR = paste(unique(IPR[!is.na(IPR)]), collapse = "; "),
    signalp = sum(signalp == "True"),
    tmhmm_geneswtmh = sum(tmhmm != "0"),
    netgpi_geneswithgpi = sum(netgpi != "Not"),
    COG = paste(unique(COG[!is.na(COG)]), collapse = "; "),
    EC = paste(unique(EC[!is.na(EC)]), collapse = "; "),
    PFAM = paste(unique(PFAM[!is.na(PFAM)]), collapse = "; ")
  )

merging_desc_sum <- merging_desc_sum %>%
  left_join(taxonomy %>% distinct(species, genus, family), by = "species")

```




```{r}
OG_df <- merging_desc_sum %>%
  mutate(sample = paste(species, strain_id, sep = "_")) %>%
  pivot_wider(names_from = sample, values_from = gene_copies_in_OG, values_fill = 0) %>%
  group_by(OG) %>%
  summarise(
    number_of_species = n_distinct(species),
    description = ifelse(
      all(description == "hypothetical protein"),
      "hypothetical protein",
      names(sort(table(description[description != "hypothetical protein"]), decreasing = TRUE))[1]
    ),
    IPR = {
      terms <- unlist(str_split(IPR, "\\s*;\\s*"))      # split all group strings
      terms <- terms[terms != "" & !is.na(terms)] 
      terms <- unique(terms)                            # keep unique
      if (length(terms) == 0) NA_character_ else paste(terms, collapse = ";")
    },
    signalp = sum(signalp, na.rm = TRUE),
    tmhmm_geneswtmh = sum(tmhmm_geneswtmh, na.rm = TRUE),
    netgpi_geneswithgpi = sum(netgpi_geneswithgpi),
    COG = {
      terms <- unlist(str_split(COG, "\\s*;\\s*"))      # split all group strings
      terms <- terms[terms != "" & !is.na(terms)] 
      terms <- unique(terms)                            # keep unique
      if (length(terms) == 0) NA_character_ else paste(terms, collapse = ";")
    },
    EC = {
      terms <- unlist(str_split(EC, "\\s*;\\s*"))      # split all group strings
      terms <- terms[terms != "" & !is.na(terms)] 
      terms <- unique(terms)                            # keep unique
      if (length(terms) == 0) NA_character_ else paste(terms, collapse = ";")
    },
    PFAM = {
      terms <- unlist(str_split(PFAM, "\\s*;\\s*"))      # split all group strings
      terms <- terms[terms != "" & !is.na(terms)] 
      terms <- unique(terms)                            # keep unique
      if (length(terms) == 0) NA_character_ else paste(terms, collapse = ";")
    },
    number_of_genera = n_distinct(genus),
    number_of_families = n_distinct(family),
    across(all_of(names(.)[15:ncol(.)]), ~ sum(.x, na.rm = TRUE)),
    .groups = "drop"
  )


```


```{r}
write_tsv(OG_df, file = "../generated/OG_annotation_genecounts.tsv")
```

```{r}
secr_df <- OG_df %>%
  filter(signalp > 0) %>%
  mutate(total_genec = rowSums(select(., 13:ncol(.)))) %>%
  select(OG,description,IPR,number_of_species,total_genec, COG, EC, PFAM, everything())
```

write to file
```{r}
write_tsv(secr_df, file = "../generated/nofilter_secr_OGxAnnos.tsv")
```



### Categorize OGs 

```{r}
secr_df <- read.csv(file = "../generated/nofilter_secr_OGxAnnos.tsv", sep = "\t")
```

#### Working on the hypothetical proteins that have IPR terms
```{r}
hypo_w_ipr <- secr_df %>% filter(description == "hypothetical protein" & IPR != "")
```

```{r}
ipr_map <- read.table("../data/ipr_entry_list.txt", sep = "\t", header = TRUE)
```

```{r}
df_new <- hypo_w_ipr %>%
  # Split multiple IPR terms into separate rows
  mutate(IPR = str_split(IPR, ";")) %>%
  unnest(IPR) %>%
  # Join to InterPro table
  left_join(ipr_map %>% select(ENTRY_AC, ENTRY_NAME),
            by = c("IPR" = "ENTRY_AC")) %>%
  # Collapse back to OG-level description
  group_by(OG) %>%
  summarise(
    ipr_description = ifelse(sum(is.na(ENTRY_NAME)) == length(ENTRY_NAME), #if all is NA = no found desc from ipr term
                         "hypothetical protein",
                         paste(na.omit(unique(ENTRY_NAME)), collapse = "; ")),
    IPR.y =  paste(na.omit(unique(IPR)), collapse = "; "),
    .groups = "drop"
  )

```


```{r}
merged <- merge(secr_df, df_new, by = "OG", all.x = TRUE)

merged <- merged %>%
  mutate(description = ifelse(description == "hypothetical protein" & ipr_description != "hypothetical protein" 
                              & !is.na(ipr_description),
                               ipr_description, 
                               description )) %>%
  select(-IPR.y, -ipr_description)

```


```{r}
write.csv(merged, file = "../generated/w_ipr_desc_nofilter_secr_OGxAnnos.tsv")
```


### Description terms  - string matching

```{r}
cat_keywords_metabolism <- list(
  "carbohydrate metabolism" = c(
    "glyco", "glucosidase", "glycosidase", "glucanase", "chitinase", "mannosidase",
    "cellulase", "saccharid", "invertase", "amylase", "xylanase", "polysaccharide",
    "beta-glucosidase", "glycosyl", "glycosyltransferase", "glycoside hydrolase", "glycosylase", "glucose",
    "chitin", "chitinase", "deacetylase", "deacetylation", "CAZy", "GH", "glycoside hydrolase", "CBM", "LysM", "glucuronidase"

  ),
  "protein metabolism" = c(
    "protease", "peptidase", "aminopeptidase", "endopeptidase", "exopeptidase",
    "serine protease", "aspartyl", "metalloprotease", "cysteine protease", "peptidyl",
    "proteinase", "proteasome", "ubiquitin", "peptidyl-prolyl", "serine/threonine", "protein kinase"
  ),
  "lipid metabolism" = c(
    "lipase", "phospholipase", "acyltransferase", "acyl", "esterase", "fatty-acid",
    "lipoxygenase", "triacylglycerol", "glycerolipid", "lipid",
    "sterol", "cholesterol", "prenyl", "isoprenoid", "phosphatidyl", "sphingolipid", "ceramide", "lanosterol", "fatty", "SUR1"))
  
```


```{r}
# Flatten synonyms mapping for canonical term extraction (map many variant tokens to canonical)
canonical_map <- list(
  "glucosidase" = c("glucosidase", "beta-glucosidase", "glycosidase", "glycoside hydrolase", "glycosylase", "glycohydrolase"),
  "chitinase" = c("chitinase"),
  "amylase" = c("amylase", "alpha-amylase"),
  "xylanase" = c("xylanase"),
  "cellulase" = c("cellulase"),
  "mannosidase" = c("mannosidase"),
  "aspartyl_protease" = c("aspartyl protease", "aspartyl", "sap", "secreted aspartyl protease"),
  "serine_protease" = c("serine protease"),
  "metalloprotease" = c("metalloprotease"),
  "peptidase" = c("peptidase", "protease", "proteinase", "aminopeptidase", "endopeptidase"),
  "lipase" = c("lipase", "phospholipase", "phospholipase b", "phospholipase a"),
  "adhesin" = c("adhesin", "agglutinin", "flocculin", "adhesion", "cell wall adhesin", "gpi-anchored adhesin"),
  "toxin" = c("toxin", "hemolysin", "invasin")
)
```

Get category function 
```{r}
get_category <- function(description, cat_keywords) {
  description <- tolower(description) 
  for (category in names(cat_keywords)) {
    patterns <- paste0(cat_keywords[[category]], collapse = "|") #Combining keywords so we can search w str detect. Faster than any()
    if (str_detect(description, patterns)) {
      return(category)
    }
  } 
  return(NA)
}
```


```{r}
merged <- merged %>% filter(number_of_species > 1 & signalp > 3)

merged <- merged %>%
  mutate(category = sapply(description, get_category, cat_keywords = cat_keywords_metabolism))

table(merged$category)
```

```{r}
get_canonical <- function(description, canonical_map) {
  description <- tolower(description)
  for (canon in names(canonical_map)) {
    patterns <- paste0(canonical_map[[canon]], collapse = "|")
    if (str_detect(description, patterns)) {
      return(canon)
    }
  }
  return(NA)
}

merged <- merged %>%
  mutate(canonical = sapply(description, get_canonical, canonical_map = canonical_map))

table(merged$canonical)

```

### Getting more categories from  COG/EC 
```{r}
cog_map <- c(
  "C" = "energy metabolism",
  "G" = "carbohydrate metabolism",
  "E" = "protein metabolism",
  # "F" = "nucleotide metabolism",
  # "H" = "coenzyme metabolism",
  "I" = "lipid metabolism" #,
  # "P" = "ion transport and metabolism",
  # "Q" = "secondary metabolite metabolism"
)

merged <- merged %>%
  mutate(category = ifelse(is.na(category) & !is.na(COG), 
              paste(unique(na.omit(cog_map[strsplit(COG, ";")[[1]]])), collapse = "; "),
              category
  )
)

table(merged$category)
```

```{r}
ec_map <- list(
  "carbohydrate metabolism" = c(
    "^1\\.1\\.", "^1\\.2\\.", "^2\\.4\\.", "^3\\.2\\.", "^4\\.2\\."  # dehydrogenases, transferases, glycosidases, lyases
  ),
  "lipid metabolism" = c(
    "^3\\.1\\.1\\.", "^3\\.1\\.4\\.", "^2\\.3\\."  # lipases, phospholipases, acyltransferases
  ),
  "protein metabolism" = c(
    "^3\\.4\\.", "^2\\.7\\."  # proteases, kinases
  ))

get_category_from_ec <- function(ec_number, ec_map) {
  if (is.na(ec_number)) return(NA)
  for (cat in names(ec_map)) {
    patterns <- ec_map[[cat]]
    if (any(str_detect(tolower(ec_number), tolower(patterns)))) {
      return(cat)
    }
  }
  return(NA)
}
```


```{r}
merged <- merged %>%
  mutate(category = ifelse(
    is.na(category) & !is.na(EC),
    sapply(EC, get_category_from_ec, ec_map = ec_map),
    category
  ))

table(merged$category)
```


We have to decide which SignalP % of genes in an OG should be considered 'secreted OGs' 
And how many different species / genera do we want? 
```{r}
merged <- merged %>% 
  filter(!tmhmm_geneswtmh >= 0.75*signalp)
```


```{r}
write_tsv(merged, file = "../generated/formanual_cur.tsv")
```

### More categories from PFAM

```{r}
test <- read_tsv("../generated/secreted_OGs_with_categories.tsv")

pfam_map <- read.delim("../generated/Pfam-A.clans.tsv", header = FALSE, sep = "\t")

# Check columns
head(pfam_map)

pfam_lookup <- pfam_map[, c(1, 5)]
colnames(pfam_lookup) <- c("Pfam_ID", "Description")

replace_pfams <- function(pfam_string, lookup) {
  # Split by commas/semicolons
  ids <- str_split(pfam_string, "[,;]")[[1]]
  # Lookup
  names <- lookup[match(ids, lookup$Pfam_ID)]$Description
  # Replace missing ones with original ID
  names[is.na(names)] <- ids[is.na(names)]
  # Recombine
  paste(unique(names), collapse = "; ")
}

# Apply to your dataframe
test$pfam_names <- map_chr(test$PFAM, replace_pfams, lookup = pfam_lookup)

```


```{r}
table(test$category)

test2 <- test %>%
  mutate(category = ifelse(is.na(category),
                           sapply(pfam_names, get_category, cat_keywords = cat_keywords_metabolism),
                           category
                           ))

table(test2$category)
```


```{r}
write_tsv(test2, file = "../generated/formanual_cur.tsv")
```



### Collapsing based on manually curated descriptions

```{r}
library(readxl)
#test_cur <- read_xlsx("../generated/test_manual_cur.xlsx")
test_cur <- read_xlsx("../generated/1211_manual_cur_secreted.xlsx")
test_cur$netgpi_geneswithgpi <- as.numeric(test_cur$netgpi_geneswithgpi)
```


We collapse OGs with the same broad descriptions and get the genecopies as sums of all gene copies. This is kept at the sample level, so for each unique description we count how many gene copies each sample has.
```{r}
collapsed_df <- test_cur %>% 
  filter(!category == "NA") %>%
  select(-c("description", "IPR", "PFAM", "EC", "COG", "canonical", "pfam_names")) %>%
  mutate(desc_collapse = str_trim(tolower(desc_collapse))) %>%
  group_by(desc_collapse) %>%
  summarise(OG = paste(unique(OG), collapse = "; "),
            category = first(category),
            number_of_species = max(number_of_species, na.rm = TRUE),
            total_genec = sum(total_genec, na.rm = TRUE),
            number_of_genera = max(number_of_genera, na.rm = TRUE),
            number_of_families = max(number_of_families, na.rm = TRUE),
            signalp = sum(as.numeric(signalp), na.rm = TRUE),
            tmhmm_geneswtmh = sum(as.numeric(tmhmm_geneswtmh), na.rm = TRUE),
            netgpi_geneswithgpi = sum(netgpi_geneswithgpi, na.rm = TRUE),
            across(all_of(names(.)[9:(ncol(.)-1)]), ~ sum(.x, na.rm = TRUE)),
            )


```


### Creating the plot dataframes


#### Count matrix 

```{r}
collapsed_filtered <- collapsed_df %>%
  filter(tmhmm_geneswtmh < 0.7*signalp) %>%
  filter(number_of_species > 2)

plot_df <- collapsed_filtered %>% select(desc_collapse, category, names(collapsed_df)[11:76])

plot_df_long <- pivot_longer(plot_df, cols = -c(desc_collapse, category), names_to = "sample", values_to = "gene_copies")

plot_df_long <- merge(plot_df_long, taxonomy[1:2], by = "sample")
length(unique(plot_df_long$sample))

#handling the hybrids
plot_df_long[plot_df_long$sample == "Malassezia_furfur_1878",]$species <- "Malassezia furfur hybrid"
plot_df_long[plot_df_long$sample == "Malassezia_furfur_6000",]$species <- "Malassezia furfur hybrid"

plot_df_long[plot_df_long$sample == "Trichosporon_ovoides_9430",]$species <- "Trichosporon ovoides hybrid"

plot_df_long[plot_df_long$sample ==  "Hortaea_werneckii_41051",]$species <- "Hortaea werneckii hybrid"
plot_df_long[plot_df_long$sample ==  "Hortaea_werneckii_100455",]$species <- "Hortaea werneckii hybrid"

plot_df_long <- plot_df_long %>%
  select(-sample) %>%
  group_by(desc_collapse, species) %>%
  summarise(gene_copies = round(mean(gene_copies)), #For the species with multiple samples, we want the average number of gene copies, but in whole numbers - the sum would put them above single-species samples. We could also use the max()
            category = first(category))

length(unique(plot_df_long$species))

plot_df_n <- pivot_wider(plot_df_long, names_from = species, values_from = gene_copies)

plot_df_n$desc_collapse <- str_to_title(plot_df_n$desc_collapse)
plot_df <- column_to_rownames(plot_df_n, "desc_collapse")

category_order <- c(
  "carbohydrate metabolism",
  "lipid metabolism",
  "protein metabolism",
  "virulence associated",
  "adhesin-associated"
)

plot_df$category <- factor(plot_df$category, levels = category_order)
plot_df <- plot_df %>% arrange(category)

```


Plotting first 

species order with hybrids
```{r}
species_order <- c("Rhodotorula toruloides", "Cryptococcus neoformans", "Trichosporon ovoides", 
                   "Trichosporon ovoides hybrid",
                   "Malassezia furfur", "Malassezia furfur hybrid", "Malassezia equina", "Malassezia dermatis", 
                   "Malassezia sympodialis", 
                   "Malassezia pachydermatis", "Malassezia restricta", "Malassezia globosa", 
                   "Pichia kudriavzevii", "Saccharomyces cerevisiae", "Nakaseomyces glabratus",
                   "Candida metapsilosis", "Candida orthopsilosis", "Candida parapsilosis", "Candida tropicalis", 
                   "Candida albicans", "Candida dubliniensis",
                   "Hortaea werneckii", "Hortaea werneckii hybrid", "Aspergillus fumigatus", "Histoplasma capsulatum", 
                   "Chrysosporium queenslandicum", 
                   "Microsporum canis", "Microsporum audouinii", "Microsporum ferrugineum", 
                   "Nannizzia gypsea", "Epidermophyton floccosum", "Trichophyton schoenleinii",
                   "Trichophyton tonsurans", "Trichophyton mentagrophytes", "Trichophyton indotineae",
                   "Trichophyton interdigitale", "Trichophyton concentricum", "Trichophyton soudanese",
                   "Trichophyton kuryangei", "Trichophyton rubrum", "Trichophyton violaceum" )

```

pallette
```{r}
hm_pallette <- rev(c("#BA2C13", "#C85A36", "#DD9F6A", "#F0E49D", "#EAEFF2", "#C7D3E0", "#A4B7CE", "#7893B7"))

article_pallette <- c("#5674b4", "#7eacd1", "#afd9e9", "#e1f2f9", "#f7e18e", "#f7b060", "#ee6d43", "#d32e28")
```


plotting the metabolism
```{r}
library(pheatmap)
subset_cats <- c("carbohydrate metabolism", "lipid metabolism", "protein metabolism")

names(plot_df)[2:42] <- sub("_", " ", names(plot_df)[2:42])

# Filter and order the dataframe
plot_df_sub <- plot_df %>%
  filter(category %in% subset_cats) %>%
  mutate(category = factor(category, levels = category_order)) %>%
  arrange(category)

# numeric matrix
num_df <- plot_df_sub[, species_order] %>%
  mutate(across(everything(), as.numeric))
mat <- as.matrix(num_df)
rownames(mat) <- rownames(plot_df_sub)

# Z scores for colour scale
mat_z <- t(scale(t(mat)))  # center and scale each row
rownames(mat_z) <- rownames(mat)
mat_z[mat_z > 2] <- 2
mat_z[mat_z < -2] <- -2

# ----- 4. Build annotation -----
annotation_row <- data.frame(Category = plot_df_sub$category)
rownames(annotation_row) <- rownames(mat)  # align annotation with matrix

# ----- 5. Compute white gaps -----
gaps_row <- cumsum(table(plot_df_sub$category))
gaps_row <- gaps_row[1:3]

# ---- Numeric labels ----
num_labels <- matrix(round(mat, 1), nrow = nrow(mat), ncol = ncol(mat))
rownames(num_labels) <- rownames(mat)
colnames(num_labels) <- colnames(mat)

# ---- Plot ----
plot <- pheatmap(mat_z,
         cluster_rows = FALSE,
         cluster_cols = FALSE,
         #annotation_row = annotation_row,
         show_rownames = TRUE,
         show_colnames = TRUE,
         fontsize_row = 10,        # size of y-axis (row) labels
         fontsize_col = 12,       # size of x-axis labels
         display_numbers = num_labels,
         number_color = "black",
         fontsize_number = 8,
         gaps_row = gaps_row,
         border_color = "white",
         color = article_pallette, 
         angle_col = 90,
         legend = FALSE,
         cellheight=16,
         cellwidth=15)

plot


rownames(mat_z)
```

give the plot bold text with grid pkg
```{r}
library(grid)
gt <- plot$gtable

# Find the column labels (x-axis labels)
col_labels_index <- which(sapply(gt$grobs, function(x) x$name) == "col_names")

# Modify the graphical parameters of species and gene product labels
gt$grobs[[2]]$gp <- gpar(fontface = "bold.italic", fontsize = 12)

gt$grobs[[3]]$gp <- gpar(fontface = "bold", fontsize = 10)
#gt$grobs[[3]]$gp <- gpar(size = 7)
grid.newpage()
grid.draw(gt)
```

save to file
```{r}
# Save plot as pdf
pdf("../plots/1311_secreted_metabolites.pdf", width = 11, height = 12, useDingbats = TRUE)  
grid.draw(gt)
dev.off()

png("../plots/1311_secreted_metabolites.png", width = 11, height = 10.5, units = "in", res = 300)  
grid.draw(gt)
dev.off()

```



```{r}
secr_df <- read.csv(file = "../generated/w_ipr_desc_nofilter_secr_OGxAnnos.tsv")
```


```{r}
gpi_df <- secr_df %>% filter(netgpi_geneswithgpi > 0) %>% select(-X)

write.csv(gpi_df, "../generated/gpi_for_manual_cur.csv")
```


## GPI anchored + virulence = figure 6


### From the secreted 

```{r}
gpi_df <- read_xlsx("../generated/gpi_manual_curated.xlsx")
gpi_df <- gpi_df %>% dplyr::select(!"Column1")

```


```{r}
collapsed_gpi <- gpi_df %>% 
  dplyr::select(-c("description", "IPR", "PFAM", "EC", "COG")) %>%
 # mutate(desc_collapse = str_trim(tolower(desc_collapse))) %>%
  group_by(desc_collapse) %>%
  summarise(OG = paste(unique(OG), collapse = "; "),
            number_of_species = max(number_of_species, na.rm = TRUE),
            total_genec = sum(total_genec, na.rm = TRUE),
            number_of_genera = max(number_of_genera, na.rm = TRUE),
            number_of_families = max(number_of_families, na.rm = TRUE),
            signalp = sum(as.numeric(signalp), na.rm = TRUE),
            tmhmm_geneswtmh = sum(as.numeric(tmhmm_geneswtmh), na.rm = TRUE),
            netgpi_geneswithgpi = sum(netgpi_geneswithgpi, na.rm = TRUE),
            across(all_of(names(.)[10:(ncol(.))]), ~ sum(.x, na.rm = TRUE)),
            )


```


```{r}
gpi_plot_df <- collapsed_gpi %>% dplyr::select(desc_collapse, names(collapsed_gpi)[10:75])

gpi_plot_df_long <- pivot_longer(gpi_plot_df, cols = -desc_collapse, names_to = "sample", values_to = "gene_copies")

gpi_plot_df_long$sample <- sub("\\.", "-", gpi_plot_df_long$sample)

gpi_plot_df_long <- merge(gpi_plot_df_long, taxonomy[1:2], by = "sample")

length(unique(gpi_plot_df_long$sample))
table(gpi_plot_df_long$sample)

#handling the hybrids
gpi_plot_df_long[gpi_plot_df_long$sample == "Malassezia_furfur_1878",]$species <- "Malassezia furfur hybrid"
gpi_plot_df_long[gpi_plot_df_long$sample == "Malassezia_furfur_6000",]$species <- "Malassezia furfur hybrid"

gpi_plot_df_long[gpi_plot_df_long$sample == "Trichosporon_ovoides_9430",]$species <- "Trichosporon ovoides hybrid"

gpi_plot_df_long[gpi_plot_df_long$sample ==  "Hortaea_werneckii_41051",]$species <- "Hortaea werneckii hybrid"
gpi_plot_df_long[gpi_plot_df_long$sample ==  "Hortaea_werneckii_100455",]$species <- "Hortaea werneckii hybrid"

gpi_plot_df_long <- gpi_plot_df_long %>%
  select(-sample) %>%
  group_by(desc_collapse, species) %>%
  summarise(gene_copies = round(mean(gene_copies))) #For the species with multiple samples, we want the average number of gene copies, but in whole numbers - the sum would put them above single-species samples. We could also use the max()

length(unique(gpi_plot_df_long$species))
length(unique(gpi_plot_df_long$desc_collapse))

gpi_plot_df_long_filtered <- gpi_plot_df_long %>%
  group_by(desc_collapse) %>%
  filter(sum(gene_copies != 0) > 1) %>%
  ungroup()

length(unique(gpi_plot_df_long_filtered$desc_collapse))

gpi_plot_df_n <- pivot_wider(gpi_plot_df_long_filtered, names_from = species, values_from = gene_copies)

#gpi_plot_df_n$desc_collapse <- str_to_title(gpi_plot_df_n$desc_collapse)

gpi_plot_df_n <- gpi_plot_df_n %>%
  arrange(desc_collapse != "GPI-anchored, unknown function", desc_collapse)

gpi_plot_df <- column_to_rownames(gpi_plot_df_n, "desc_collapse")

```


species order with hybrids
```{r}
species_order_hybrids <- c("Rhodotorula toruloides", "Cryptococcus neoformans", "Trichosporon ovoides", 
                   "Trichosporon ovoides hybrid",
                   "Malassezia furfur", "Malassezia furfur hybrid", "Malassezia equina", "Malassezia dermatis", 
                   "Malassezia sympodialis", 
                   "Malassezia pachydermatis", "Malassezia restricta", "Malassezia globosa", 
                   "Pichia kudriavzevii", "Saccharomyces cerevisiae", "Nakaseomyces glabratus",
                   "Candida metapsilosis", "Candida orthopsilosis", "Candida parapsilosis", "Candida tropicalis", 
                   "Candida albicans", "Candida dubliniensis",
                   "Hortaea werneckii", "Hortaea werneckii hybrid", "Aspergillus fumigatus", "Histoplasma capsulatum", 
                   "Chrysosporium queenslandicum", 
                   "Microsporum canis", "Microsporum audouinii", "Microsporum ferrugineum", 
                   "Nannizzia gypsea", "Epidermophyton floccosum", "Trichophyton schoenleinii",
                   "Trichophyton tonsurans", "Trichophyton mentagrophytes", "Trichophyton indotineae",
                   "Trichophyton interdigitale", "Trichophyton concentricum", "Trichophyton soudanese",
                   "Trichophyton kuryangei", "Trichophyton rubrum", "Trichophyton violaceum" )

```


plotting secreted gpi
```{r}
names(gpi_plot_df)[1:41] <- sub("_", " ", names(gpi_plot_df)[1:41])

# numeric matrix
num_df <- gpi_plot_df[, species_order_hybrids] %>%
  mutate(across(everything(), as.numeric))
mat <- as.matrix(num_df)
rownames(mat) <- rownames(gpi_plot_df)

rns_ordered <- c(setdiff(sort(rownames(mat)), "GPI-anchored, unknown function"),
                 "GPI-anchored, unknown function")

# reorder matrix rows
mat <- mat[rns_ordered, ]


# Z scores for colour scale
mat_z <- t(scale(t(mat)))  # center and scale each row
rownames(mat_z) <- rownames(mat)
mat_z[mat_z > 2] <- 2
mat_z[mat_z < -2] <- -2

# ----- 4. Build annotation -----
# annotation_row <- data.frame(Category = gpi_plot_df$category)
# rownames(annotation_row) <- rownames(mat)  # align annotation with matrix
# 
# # ----- 5. Compute white gaps -----
# gaps_row <- cumsum(table(plot_df_sub$category))

# ---- Numeric labels ----
num_labels <- matrix(round(mat, 1), nrow = nrow(mat), ncol = ncol(mat))
rownames(num_labels) <- rownames(mat)
colnames(num_labels) <- colnames(mat)

# ---- Plot ----
library(pheatmap)
plot <- pheatmap(mat_z,
         cluster_rows = FALSE,
         cluster_cols = FALSE,
         #annotation_row = annotation_row,
         show_rownames = TRUE,
         show_colnames = TRUE,
         fontsize_row = 7,        # size of y-axis (row) labels
         fontsize_col = 10,       # size of x-axis labels
         display_numbers = num_labels,
         number_color = "black",
         fontsize_number = 7,
        # gaps_row = gaps_row,
        # border_color = "white",
         color = article_pallette, 
         angle_col = 90,
         legend = FALSE)

plot

```

give the plot bold text with grid pkg
```{r}
library(grid)
gt <- plot$gtable

# Find the column labels (x-axis labels)
col_labels_index <- which(sapply(gt$grobs, function(x) x$name) == "col_names")

# Modify the graphical parameters of species and gene product labels
gt$grobs[[2]]$gp <- gpar(fontface = "bold.italic") #species labels 

gt$grobs[[3]]$gp <- gpar(fontface = "bold", fontsize = 8) #descriptions

grid.newpage()
grid.draw(gt)

```

save to file
```{r}
# Save plot as pdf
pdf("../plots/0511_gpi_anchored.pdf", width = 10, height = 10, useDingbats = TRUE)  
grid.draw(gt)
dev.off()

png("../plots/0411_secreted_metabolites.png", width = 10, height = 10, units = "in", res = 300)  
grid.draw(gt)
dev.off()

```

### From the non-secreted - getting adhesins and virulence-associated 

Just to see if we find descriptions that are belived to be adhesins / GPI anchored, but where not predicted secreted / GPI-anchored by the prediction tools. Here we can ofc only look for well-described adhesin genes. 



Adhesins 
```{r}
OGs_x_annos <- OGs_x_annos %>%
  filter(!OG %in% gpi_df$OG)

```


```{r}
cat_keywords_viru <- list( 
 "virulence associated" = c(
    "toxin", "virulence", "pathogen", "pathogenic", "invasin", "virulence factor",
    "hemolysin", "phospholipase b", "secreted aspartyl protease", "secreted protease",
    "lysM", "Hydrophobin", "Heat shock", "CFEM", "BNR repeat", "PIR", "necrosis", "stress", "igE"
  ),
  "adhesin-associated" = c(
    "adhesin", "adhesion", "agglutinin", "flocculin", "intimin", "fimbrial", "adhesive",
    "gpi-anchored adhesin", "gpi-anchored", "cell wall adhesin", "Flo11", "FLO8", "GLEYA",
    "^PA14", "Hyphal_reg", "^Hyr", "GPI", "Glycosylphosphatidylinositol"
    
  )
)
```


```{r}
virulence_assoc <- OGs_x_annos %>%
  mutate(category = sapply(description, get_category, cat_keywords = cat_keywords_viru)) %>%
  mutate(category = ifelse(
    is.na(category),
    sapply(pfam_names, get_category, cat_keywords = cat_keywords_viru),
    category
  )) %>%
  filter(!is.na(category))

table(virulence_assoc$category)

virulence_assoc <- virulence_assoc %>% filter(!OG %in% comb_gpi$OG) 

#write.csv(virulence_assoc, "../generated/viru_assoc_formancur.csv")
```

### Start here from the 1211_manual_cur excel sheet
adhesin and virulence associated from manually curated excel sheets. Both secreted and non-secreted 

species order with hybrids
```{r}
species_order_hybrids <- c("Rhodotorula toruloides", "Cryptococcus neoformans", "Trichosporon ovoides", 
                   "Trichosporon ovoides hybrid",
                   "Malassezia furfur", "Malassezia furfur hybrid", "Malassezia equina", "Malassezia dermatis", 
                   "Malassezia sympodialis", 
                   "Malassezia pachydermatis", "Malassezia restricta", "Malassezia globosa", 
                   "Pichia kudriavzevii", "Saccharomyces cerevisiae", "Nakaseomyces glabratus",
                   "Candida metapsilosis", "Candida orthopsilosis", "Candida parapsilosis", "Candida tropicalis", 
                   "Candida albicans", "Candida dubliniensis",
                   "Hortaea werneckii", "Hortaea werneckii hybrid", "Aspergillus fumigatus", "Histoplasma capsulatum", 
                   "Chrysosporium queenslandicum", 
                   "Microsporum canis", "Microsporum audouinii", "Microsporum ferrugineum", 
                   "Nannizzia gypsea", "Epidermophyton floccosum", "Trichophyton schoenleinii",
                   "Trichophyton tonsurans", "Trichophyton mentagrophytes", "Trichophyton indotineae",
                   "Trichophyton interdigitale", "Trichophyton concentricum", "Trichophyton soudanese",
                   "Trichophyton kuryangei", "Trichophyton rubrum", "Trichophyton violaceum" )

```


```{r}
test_cur <- read_xlsx("../generated/1211_manual_cur_secreted.xlsx")
virulence_assoc <- test_cur %>% filter(category == "virulence associated"| category == "adhesin-associated") %>%
  dplyr::select(-c("description", "IPR", "PFAM", "EC", "COG", "canonical"))

viru_non_secr <- read_xlsx("../generated/1211_manual_cur_secreted.xlsx", sheet = 2) %>%
  filter(!OG %in% test_cur$OG) %>% dplyr::select(-c("description","IPR", "PFAM", "EC", "COG"))

names(viru_non_secr)   %in% names(virulence_assoc)
names(virulence_assoc) %in% names(viru_non_secr)

names(viru_non_secr)[9:75] <- sub("\\.", "-", names(viru_non_secr)[9:75])

viru_non_secr$total_genec <- rowSums(viru_non_secr[9:74] != 0)

names(viru_non_secr)   %in% names(virulence_assoc)
names(virulence_assoc) %in% names(viru_non_secr)

viru_non_secr$secr <- FALSE
virulence_assoc$secr <- TRUE

comb_viru <- rbind(viru_non_secr, virulence_assoc)

table(comb_viru$category)
table(comb_viru$secr)
dim(comb_viru)
```


```{r}
#write.csv(viru_non_secr, "../generated/viru_non_secr.csv")
#test <- read_csv("../generated/viru_non_secr.csv")

#test$OG %in% test_cur$OG
```


```{r}
collapsed_gpi_viru <- comb_viru %>% 
  group_by(desc_collapse) %>%
  dplyr::select(-pfam_names) %>%
  filter(!is.na(category)) %>%
  summarise(OG = paste(unique(OG), collapse = "; "),
            number_of_species = max(number_of_species, na.rm = TRUE),
            total_genec = sum(total_genec, na.rm = TRUE),
            number_of_genera = max(number_of_genera, na.rm = TRUE),
            number_of_families = max(number_of_families, na.rm = TRUE),
            signalp = sum(as.numeric(signalp), na.rm = TRUE),
            tmhmm_geneswtmh = sum(as.numeric(tmhmm_geneswtmh), na.rm = TRUE),
            netgpi_geneswithgpi = sum(as.numeric(netgpi_geneswithgpi), na.rm = TRUE),
            category = first(category),
            secr = if (length(unique(secr)) == 1) as.character(unique(secr)) else "mixed",
            across(all_of(names(.)[9:(ncol(.)-3)]), ~ sum(.x, na.rm = TRUE)),
            )

```


```{r}
gpi_plot_df <- collapsed_gpi_viru %>% dplyr::select(desc_collapse, category, secr, netgpi_geneswithgpi, names(collapsed_gpi_viru)[12:77]) %>% filter(!is.na(category))

gpi_plot_df_long <- pivot_longer(gpi_plot_df, cols = -c(desc_collapse, category, secr, netgpi_geneswithgpi), names_to = "sample", values_to = "gene_copies")

gpi_plot_df_long <- merge(gpi_plot_df_long, taxonomy[1:2], by = "sample")

#write.csv(gpi_plot_df_long, "../generated/gpi_plot_df_long.csv")

length(unique(gpi_plot_df_long$sample))
table(gpi_plot_df_long$sample)

#handling the hybrids
gpi_plot_df_long[gpi_plot_df_long$sample == "Malassezia_furfur_1878",]$species <- "Malassezia furfur hybrid"
gpi_plot_df_long[gpi_plot_df_long$sample == "Malassezia_furfur_6000",]$species <- "Malassezia furfur hybrid"

gpi_plot_df_long[gpi_plot_df_long$sample == "Trichosporon_ovoides_9430",]$species <- "Trichosporon ovoides hybrid"

gpi_plot_df_long[gpi_plot_df_long$sample ==  "Hortaea_werneckii_41051",]$species <- "Hortaea werneckii hybrid"
gpi_plot_df_long[gpi_plot_df_long$sample ==  "Hortaea_werneckii_100455",]$species <- "Hortaea werneckii hybrid"

gpi_plot_df_long <- gpi_plot_df_long %>%
  select(-sample) %>%
  group_by(desc_collapse, species) %>%
  summarise(gene_copies = round(mean(gene_copies)),#For the species with multiple samples, we want the average number of gene copies, but in whole numbers - the sum would put them above single-species samples. We could also use the max()
            category = first(category),
            secr = first(secr),
            GPI = first(netgpi_geneswithgpi)) 

length(unique(gpi_plot_df_long$species))
length(unique(gpi_plot_df_long$desc_collapse))

gpi_plot_df_long_filtered <- gpi_plot_df_long %>%
  group_by(desc_collapse) %>%
  filter(sum(gene_copies != 0) > 1) %>%
  ungroup()

length(unique(gpi_plot_df_long_filtered$desc_collapse))

gpi_plot_df_n <- pivot_wider(gpi_plot_df_long_filtered, names_from = species, values_from = gene_copies)

gpi_plot_df_n <- gpi_plot_df_n[order(
  gpi_plot_df_n$category,
  gpi_plot_df_n$desc_collapse == "GPI-anchored, unannotated"
), ]

gpi_plot_df <- column_to_rownames(gpi_plot_df_n, "desc_collapse")

```


plotting 
```{r}
names(gpi_plot_df)[2:44] <- sub("_", " ", names(gpi_plot_df)[2:44])

# numeric matrix
num_df <- gpi_plot_df[, species_order_hybrids] %>%
  mutate(across(everything(), as.numeric))
mat <- as.matrix(num_df)
rownames(mat) <- rownames(gpi_plot_df)

# rns_ordered <- c(
#   setdiff(rownames(gpi_plot_df)[order(gpi_plot_df$category)], "GPI-anchored, unknown function"),
#   "GPI-anchored, unknown function"
# )
# # reorder matrix rows
# mat <- mat[rns_ordered, ]

# Z scores for colour scale
mat_z <- t(scale(t(mat)))  # center and scale each row
rownames(mat_z) <- rownames(mat)
mat_z[mat_z > 2] <- 2
mat_z[mat_z < -2] <- -2

# ----- 4. Build annotation -----
annotation_row <- data.frame(GPI = as.character(gpi_plot_df$GPI>0), secreted = gpi_plot_df$secr)
rownames(annotation_row) <-  (rownames(mat))  # align annotation with matrix


# # ----- 5. Compute white gaps -----
gaps_row <- cumsum(table(gpi_plot_df$category))

# ---- Numeric labels ----
num_labels <- matrix(round(mat, 1), nrow = nrow(mat), ncol = ncol(mat))
rownames(num_labels) <- rownames(mat)
colnames(num_labels) <- colnames(mat)

# ---- Plot ----
library(pheatmap)
plot <- pheatmap(mat_z,
         cluster_rows = FALSE,
         cluster_cols = FALSE,
         annotation_row = annotation_row,
         show_rownames = TRUE,
         show_colnames = TRUE,
         fontsize_row = 8,        # size of y-axis (row) labels
         fontsize_col = 10,       # size of x-axis labels
         display_numbers = num_labels,
         number_color = "black",
         fontsize_number = 8,
         gaps_row = gaps_row,
         border_color = "white",
         color = article_pallette, 
         angle_col = 90,
         legend = FALSE,
        # annotation_legend = FALSE,
       cellheight=10,
       cellwidth=13)

plot

```
cellheight=16,
cellwidth=15)

give the plot bold text with grid pkg
```{r}
library(grid)
gt <- plot$gtable

# Find the column labels (x-axis labels)
col_labels_index <- which(sapply(gt$grobs, function(x) x$name) == "col_names")

# Modify the graphical parameters of species and gene product labels
gt$grobs[[2]]$gp <- gpar(fontface = "bold.italic", fontsize = 10) #species labels 

gt$grobs[[3]]$gp <- gpar(fontface = "bold", fontsize = 9) #descriptions

grid.newpage()
grid.draw(gt)

```

save to file
```{r}
# Save plot as pdf
pdf("../plots/2111_gpi_viru_hm.pdf", width = 10.5, height = 11, useDingbats = TRUE)  
grid.draw(gt)
dev.off()

png("../plots/2111_gpi_viru_hm.png", width = 10.5, height = 11, units = "in", res = 300)  
grid.draw(gt)
dev.off()

```



#### making supplementary tables

```{r}
OGs_x_annos <- read.csv("../generated/OG_annotation_genecounts.tsv", sep = "\t")

```


```{r}
# From UniProt (current as of 2025)
pfam_map <- read.delim("../generated/Pfam-A.clans.tsv", header = FALSE, sep = "\t")

# Check columns
head(pfam_map)

pfam_lookup <- pfam_map[, c(1, 5)]
colnames(pfam_lookup) <- c("Pfam_ID", "Description")

replace_pfams <- function(pfam_string, lookup) {
  # Split by commas/semicolons
  ids <- str_split(pfam_string, "[,;]")[[1]]
  # Lookup
  names <- lookup$Description[match(ids, lookup$Pfam_ID)]
  # Replace missing ones with original ID
  names[is.na(names)] <- ids[is.na(names)]
  # Recombine
  paste(unique(names), collapse = "; ")
}

# Give pfam names 
OGs_x_annos$pfam_names <- map_chr(OGs_x_annos$PFAM, replace_pfams, lookup = pfam_lookup)

OGs_x_annos$PFAM


```


```{r}
cat_keywords <- list( 
  "carbohydrate metabolism" = c(
    "glyco", "glucosidase", "glycosidase", "glucanase", "chitinase", "mannosidase",
    "cellulase", "saccharid", "invertase", "amylase", "xylanase", "polysaccharide",
    "beta-glucosidase", "glycosyl", "glycosyltransferase", "glycoside hydrolase", "glycosylase", "glucose",
    "chitin", "chitinase", "deacetylase", "deacetylation", "CAZy", "GH", "glycoside hydrolase", "CBM", "LysM", "glucuronidase"

  ),
  "protein metabolism" = c(
    "protease", "peptidase", "aminopeptidase", "endopeptidase", "exopeptidase",
    "serine protease", "aspartyl", "metalloprotease", "cysteine protease", "peptidyl",
    "proteinase", "proteasome", "ubiquitin", "peptidyl-prolyl", "serine/threonine", "protein kinase"
  ),
  "lipid metabolism" = c(
    "lipase", "phospholipase", "acyltransferase", "acyl", "esterase", "fatty-acid",
    "lipoxygenase", "triacylglycerol", "glycerolipid", "lipid",
    "sterol", "cholesterol", "prenyl", "isoprenoid", "phosphatidyl", "sphingolipid", "ceramide", "lanosterol", "fatty", "SUR1"
    ),
 "virulence associated" = c(
    "toxin", "virulence", "pathogen", "pathogenic", "invasin", "virulence factor",
    "hemolysin", "phospholipase b", "secreted aspartyl protease", "sap", "secreted protease",
    "lysM", "Hydrophobin", "Heat shock", "CFEM", "BNR repeat", "PIR", "necrosis", "stress", "igE"
  ),
  "adhesin-associated" = c(
    "adhesin", "adhesion", "agglutinin", "flocculin", "intimin", "fimbrial", "adhesive",
    "gpi-anchored adhesin", "gpi-anchored", "cell wall adhesin", "Flo11", "FLO8", "GLEYA",
    "^PA14", "Hyphal_reg", "^Hyr", "GPI", "Glycosylphosphatidylinositol"
    
  )
)
```

```{r}
get_category <- function(description, cat_keywords) {
  description <- tolower(description) 
  for (category in names(cat_keywords)) {
    patterns <- paste0(cat_keywords[[category]], collapse = "|") #Combining keywords so we can search w str detect. Faster than any()
    if (str_detect(description, patterns)) {
      return(category)
    }
  } 
  return(NA)
}
```


```{r}
OGs_x_annos <- OGs_x_annos %>%
  mutate(category = sapply(description, get_category, cat_keywords = cat_keywords)) %>%
  mutate(category = ifelse(
    is.na(category),
    sapply(pfam_names, get_category, cat_keywords = cat_keywords),
    category
  ))

table(OGs_x_annos$category)
```


```{r}
write_tsv(OGs_x_annos, "../generated/OGs_x_anno_incl_pfamnames.tsv")
```

